{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Expression Data Import\n",
    "Author: Catia Antunes   \n",
    "Date: 04/03/2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import requests\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get the data\n",
    "\n",
    "def get_geo_data(gene_list):\n",
    "    # Join the list of genes with '+' for the API query\n",
    "    genes_query = '+'.join(gene_list)\n",
    "\n",
    "    # Construct the GEO API URL with your gene list\n",
    "    geo_api_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&term={genes_query}&retmode=json\"  # change this line\n",
    "\n",
    "    # Make a request to GEO API\n",
    "    response = requests.get(geo_api_url)\n",
    "\n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        # Extract the gene IDs from the API response\n",
    "        gene_ids = response.json()['esearchresult']['idlist']\n",
    "\n",
    "        # Construct the GEO query URL for expression data\n",
    "        geo_query_url = f\"https://www.ncbi.nlm.nih.gov/geo/series/{gene_ids[0]}/gds{gene_ids[0]}/\"\n",
    "\n",
    "        # Make a request to the GEO query URL\n",
    "        geo_response = requests.get(geo_query_url)\n",
    "\n",
    "        # Check if the request was successful\n",
    "        if geo_response.status_code == 200:\n",
    "            # Extract and print the relevant information from the GEO page\n",
    "            print(\"Gene Expression Data:\")\n",
    "            print(geo_response.text)\n",
    "        else:\n",
    "            print(f\"Error accessing GEO data: {geo_response.status_code}\")\n",
    "    else:\n",
    "        print(f\"Error accessing GEO API: {response.status_code}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>143</td>\n",
       "      <td>cg11728741</td>\n",
       "      <td>0.928874</td>\n",
       "      <td>8</td>\n",
       "      <td>41896469</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>286</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>ANK1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>8.0</td>\n",
       "      <td>22975146.0</td>\n",
       "      <td>Ank1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>297</td>\n",
       "      <td>cg24352905</td>\n",
       "      <td>0.853268</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13.0</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9.0</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>160</td>\n",
       "      <td>cg13058338</td>\n",
       "      <td>0.638660</td>\n",
       "      <td>12</td>\n",
       "      <td>54173598</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>23583</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SMUG1_Intron</td>\n",
       "      <td>Intron (ENST00000635234.1/23583, intron 2 of 4)</td>\n",
       "      <td>15.0</td>\n",
       "      <td>103146996.0</td>\n",
       "      <td>Smug1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>316</td>\n",
       "      <td>cg26067250</td>\n",
       "      <td>0.510613</td>\n",
       "      <td>2</td>\n",
       "      <td>172085721</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>1745</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>DLX1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>71530037.0</td>\n",
       "      <td>Dlx1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock1  CHR    bp_hg38       Gene Gene.hg19  \\\n",
       "0    143  cg11728741     0.928874    8   41896469       ANK1      ANK1   \n",
       "1    297  cg24352905     0.853268    5   77645452        OTP       OTP   \n",
       "2     98  cg08938156     0.843749    3  147409417  LOC440982      ZIC1   \n",
       "3    160  cg13058338     0.638660   12   54173598      SMUG1     SMUG1   \n",
       "4    316  cg26067250     0.510613    2  172085721       DLX1      DLX1   \n",
       "\n",
       "   ENTREZID                           conservationInMouse       GeneRegionID  \\\n",
       "0       286  conserved gene and region in mouse and human          ANK1_Exon   \n",
       "1     23440  conserved gene and region in mouse and human       OTP_Promoter   \n",
       "2    440982   mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "3     23583   mapped to different gene in human and mouse       SMUG1_Intron   \n",
       "4      1745   mapped to different gene in human and mouse          DLX1_Exon   \n",
       "\n",
       "                                        annotation  CHR_mm10      bp_mm10  \\\n",
       "0                                 Promoter (<=1kb)       8.0   22975146.0   \n",
       "1                                 Promoter (5-6kb)      13.0   94869014.0   \n",
       "2                                           5' UTR       9.0   91365746.0   \n",
       "3  Intron (ENST00000635234.1/23583, intron 2 of 4)      15.0  103146996.0   \n",
       "4                                 Promoter (<=1kb)       2.0   71530037.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0      Ank1  \n",
       "1       Otp  \n",
       "2      Zic1  \n",
       "3     Smug1  \n",
       "4      Dlx1  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Clock 1\n",
    "Clock1_top25 = pd.read_csv('../../FunctionalAnalysis/top_25_cpgs_clock_1.csv')\n",
    "Clock1_top25.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock2</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>142</td>\n",
       "      <td>cg05474883</td>\n",
       "      <td>0.952198</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>5.0</td>\n",
       "      <td>99962129.0</td>\n",
       "      <td>Hnrnpd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>272</td>\n",
       "      <td>cg09710440</td>\n",
       "      <td>0.877027</td>\n",
       "      <td>7</td>\n",
       "      <td>104328932</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>375612</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>LHFPL3_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>783</td>\n",
       "      <td>cg26512254</td>\n",
       "      <td>0.693060</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOTTIP</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>Intron (ENST00000421733.1/100316868, intron 1 ...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>52260841.0</td>\n",
       "      <td>Hoxa13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>238</td>\n",
       "      <td>cg08681110</td>\n",
       "      <td>0.647719</td>\n",
       "      <td>6</td>\n",
       "      <td>113857322</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>4082</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>MARCKS_Promoter</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>10.0</td>\n",
       "      <td>37138948.0</td>\n",
       "      <td>Marcks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>253</td>\n",
       "      <td>cg09227056</td>\n",
       "      <td>0.615235</td>\n",
       "      <td>2</td>\n",
       "      <td>176075721</td>\n",
       "      <td>EVX2</td>\n",
       "      <td>EVX2</td>\n",
       "      <td>344191</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>EVX2_Intergenic_downstream</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>74651253.0</td>\n",
       "      <td>Evx2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock2  CHR    bp_hg38    Gene Gene.hg19  \\\n",
       "0    142  cg05474883     0.952198    4   82354926  HNRNPD    HNRNPD   \n",
       "1    272  cg09710440     0.877027    7  104328932  LHFPL3    LHFPL3   \n",
       "2    783  cg26512254     0.693060    7   27200144  HOTTIP    HOXA13   \n",
       "3    238  cg08681110     0.647719    6  113857322  MARCKS    MARCKS   \n",
       "4    253  cg09227056     0.615235    2  176075721    EVX2      EVX2   \n",
       "\n",
       "    ENTREZID                           conservationInMouse  \\\n",
       "0       3184  conserved gene and region in mouse and human   \n",
       "1     375612          mapped in human, not mapped in mouse   \n",
       "2  100316868   mapped to different gene in human and mouse   \n",
       "3       4082  conserved gene and region in mouse and human   \n",
       "4     344191  conserved gene and region in mouse and human   \n",
       "\n",
       "                 GeneRegionID  \\\n",
       "0             HNRNPD_threeUTR   \n",
       "1                 LHFPL3_Exon   \n",
       "2               HOTTIP_Intron   \n",
       "3             MARCKS_Promoter   \n",
       "4  EVX2_Intergenic_downstream   \n",
       "\n",
       "                                          annotation  CHR_mm10     bp_mm10  \\\n",
       "0                                             3' UTR       5.0  99962129.0   \n",
       "1                                   Promoter (<=1kb)       NaN         NaN   \n",
       "2  Intron (ENST00000421733.1/100316868, intron 1 ...       6.0  52260841.0   \n",
       "3                                   Promoter (<=1kb)      10.0  37138948.0   \n",
       "4                                  Distal Intergenic       2.0  74651253.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0    Hnrnpd  \n",
       "1       NaN  \n",
       "2    Hoxa13  \n",
       "3    Marcks  \n",
       "4      Evx2  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Clock 2\n",
    "Clock2_top25 = pd.read_csv('../../FunctionalAnalysis/top_25_cpgs_clock_2.csv')\n",
    "Clock2_top25.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock3</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>493</td>\n",
       "      <td>cg26512254</td>\n",
       "      <td>2.638299</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOTTIP</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>Intron (ENST00000421733.1/100316868, intron 1 ...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>52260841.0</td>\n",
       "      <td>Hoxa13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>99</td>\n",
       "      <td>cg09710440</td>\n",
       "      <td>1.854689</td>\n",
       "      <td>7</td>\n",
       "      <td>104328932</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>375612</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>LHFPL3_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34</td>\n",
       "      <td>cg18418719</td>\n",
       "      <td>1.806028</td>\n",
       "      <td>3</td>\n",
       "      <td>70972083</td>\n",
       "      <td>FOXP1</td>\n",
       "      <td>FOXP1</td>\n",
       "      <td>27086</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>FOXP1_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>6.0</td>\n",
       "      <td>98941067.0</td>\n",
       "      <td>Foxp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>134</td>\n",
       "      <td>cg24352905</td>\n",
       "      <td>1.738218</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13.0</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>576</td>\n",
       "      <td>cg05474883</td>\n",
       "      <td>1.463495</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>5.0</td>\n",
       "      <td>99962129.0</td>\n",
       "      <td>Hnrnpd</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock3  CHR    bp_hg38    Gene Gene.hg19  \\\n",
       "0    493  cg26512254     2.638299    7   27200144  HOTTIP    HOXA13   \n",
       "1     99  cg09710440     1.854689    7  104328932  LHFPL3    LHFPL3   \n",
       "2     34  cg18418719     1.806028    3   70972083   FOXP1     FOXP1   \n",
       "3    134  cg24352905     1.738218    5   77645452     OTP       OTP   \n",
       "4    576  cg05474883     1.463495    4   82354926  HNRNPD    HNRNPD   \n",
       "\n",
       "    ENTREZID                           conservationInMouse     GeneRegionID  \\\n",
       "0  100316868   mapped to different gene in human and mouse    HOTTIP_Intron   \n",
       "1     375612          mapped in human, not mapped in mouse      LHFPL3_Exon   \n",
       "2      27086   mapped to different gene in human and mouse   FOXP1_threeUTR   \n",
       "3      23440  conserved gene and region in mouse and human     OTP_Promoter   \n",
       "4       3184  conserved gene and region in mouse and human  HNRNPD_threeUTR   \n",
       "\n",
       "                                          annotation  CHR_mm10     bp_mm10  \\\n",
       "0  Intron (ENST00000421733.1/100316868, intron 1 ...       6.0  52260841.0   \n",
       "1                                   Promoter (<=1kb)       NaN         NaN   \n",
       "2                                             3' UTR       6.0  98941067.0   \n",
       "3                                   Promoter (5-6kb)      13.0  94869014.0   \n",
       "4                                             3' UTR       5.0  99962129.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0    Hoxa13  \n",
       "1       NaN  \n",
       "2     Foxp1  \n",
       "3       Otp  \n",
       "4    Hnrnpd  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Clock 3\n",
    "Clock3_top25 = pd.read_csv('../../FunctionalAnalysis/top_25_cpgs_clock_3.csv')\n",
    "Clock3_top25.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock2</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR_y</th>\n",
       "      <th>bp_hg38_y</th>\n",
       "      <th>Gene.hg19_y</th>\n",
       "      <th>ENTREZID_y</th>\n",
       "      <th>conservationInMouse_y</th>\n",
       "      <th>GeneRegionID_y</th>\n",
       "      <th>annotation_y</th>\n",
       "      <th>CHR_mm10_y</th>\n",
       "      <th>bp_mm10_y</th>\n",
       "      <th>Gene_mm10_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>142</td>\n",
       "      <td>cg05474883</td>\n",
       "      <td>0.952198</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>5</td>\n",
       "      <td>99962129.0</td>\n",
       "      <td>Hnrnpd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>783</td>\n",
       "      <td>cg26512254</td>\n",
       "      <td>0.693060</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOTTIP</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>Intron (ENST00000421733.1/100316868, intron 1 ...</td>\n",
       "      <td>6</td>\n",
       "      <td>52260841.0</td>\n",
       "      <td>Hoxa13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>238</td>\n",
       "      <td>cg08681110</td>\n",
       "      <td>0.647719</td>\n",
       "      <td>6</td>\n",
       "      <td>113857322</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>4082</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>MARCKS_Promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>113857322</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>4082</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>MARCKS_Promoter</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>10</td>\n",
       "      <td>37138948.0</td>\n",
       "      <td>Marcks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>803</td>\n",
       "      <td>cg27201382</td>\n",
       "      <td>0.523618</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>109676257.0</td>\n",
       "      <td>Bdnf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>599</td>\n",
       "      <td>cg19927064</td>\n",
       "      <td>0.499713</td>\n",
       "      <td>12</td>\n",
       "      <td>48818919</td>\n",
       "      <td>CACNB3</td>\n",
       "      <td>CACNB3</td>\n",
       "      <td>784</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>CACNB3_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>48818919</td>\n",
       "      <td>CACNB3</td>\n",
       "      <td>784</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>CACNB3_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>15</td>\n",
       "      <td>98635158.0</td>\n",
       "      <td>Cacnb3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_x       var_x  beta_clock2  CHR_x  bp_hg38_x    Gene Gene.hg19_x  \\\n",
       "0      142  cg05474883     0.952198      4   82354926  HNRNPD      HNRNPD   \n",
       "1      783  cg26512254     0.693060      7   27200144  HOTTIP      HOXA13   \n",
       "2      238  cg08681110     0.647719      6  113857322  MARCKS      MARCKS   \n",
       "3      803  cg27201382     0.523618     11   27720483    BDNF        BDNF   \n",
       "4      599  cg19927064     0.499713     12   48818919  CACNB3      CACNB3   \n",
       "\n",
       "   ENTREZID_x                         conservationInMouse_x   GeneRegionID_x  \\\n",
       "0        3184  conserved gene and region in mouse and human  HNRNPD_threeUTR   \n",
       "1   100316868   mapped to different gene in human and mouse    HOTTIP_Intron   \n",
       "2        4082  conserved gene and region in mouse and human  MARCKS_Promoter   \n",
       "3         627  conserved gene and region in mouse and human     BDNF_fiveUTR   \n",
       "4         784  conserved gene and region in mouse and human   CACNB3_fiveUTR   \n",
       "\n",
       "   ... CHR_y  bp_hg38_y  Gene.hg19_y ENTREZID_y  \\\n",
       "0  ...     4   82354926       HNRNPD       3184   \n",
       "1  ...     7   27200144       HOXA13  100316868   \n",
       "2  ...     6  113857322       MARCKS       4082   \n",
       "3  ...    11   27720483         BDNF        627   \n",
       "4  ...    12   48818919       CACNB3        784   \n",
       "\n",
       "                          conservationInMouse_y   GeneRegionID_y  \\\n",
       "0  conserved gene and region in mouse and human  HNRNPD_threeUTR   \n",
       "1   mapped to different gene in human and mouse    HOTTIP_Intron   \n",
       "2  conserved gene and region in mouse and human  MARCKS_Promoter   \n",
       "3  conserved gene and region in mouse and human     BDNF_fiveUTR   \n",
       "4  conserved gene and region in mouse and human   CACNB3_fiveUTR   \n",
       "\n",
       "                                        annotation_y  CHR_mm10_y    bp_mm10_y  \\\n",
       "0                                             3' UTR           5   99962129.0   \n",
       "1  Intron (ENST00000421733.1/100316868, intron 1 ...           6   52260841.0   \n",
       "2                                   Promoter (<=1kb)          10   37138948.0   \n",
       "3                                   Promoter (<=1kb)           2  109676257.0   \n",
       "4                                             5' UTR          15   98635158.0   \n",
       "\n",
       "  Gene_mm10_y  \n",
       "0      Hnrnpd  \n",
       "1      Hoxa13  \n",
       "2      Marcks  \n",
       "3        Bdnf  \n",
       "4      Cacnb3  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Clock overlap 2 and 3\n",
    "Clock_overlap_2_3_top25 = pd.read_csv('../../FunctionalAnalysis/top_25_cpgs_overlap_2_3.csv')\n",
    "Clock_overlap_2_3_top25.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9.0</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>329</td>\n",
       "      <td>cg27201382</td>\n",
       "      <td>0.444645</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>109676257.0</td>\n",
       "      <td>Bdnf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>144</td>\n",
       "      <td>cg11904056</td>\n",
       "      <td>0.341791</td>\n",
       "      <td>5</td>\n",
       "      <td>93583204</td>\n",
       "      <td>NR2F1-AS1</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>93585317</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>13.0</td>\n",
       "      <td>78198288.0</td>\n",
       "      <td>Nr2f1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>44</td>\n",
       "      <td>cg03820088</td>\n",
       "      <td>0.337644</td>\n",
       "      <td>7</td>\n",
       "      <td>23522229</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>29896</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>TRA2A_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>23522229</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>29896</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>TRA2A_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>6.0</td>\n",
       "      <td>49252905.0</td>\n",
       "      <td>Tra2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>58</td>\n",
       "      <td>cg04998737</td>\n",
       "      <td>0.336275</td>\n",
       "      <td>5</td>\n",
       "      <td>77645403</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13.0</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_x       var_x  beta_clock1  CHR_x  bp_hg38_x       Gene Gene.hg19_x  \\\n",
       "0       98  cg08938156     0.843749      3  147409417  LOC440982        ZIC1   \n",
       "1      329  cg27201382     0.444645     11   27720483       BDNF        BDNF   \n",
       "2      144  cg11904056     0.341791      5   93583204  NR2F1-AS1       NR2F1   \n",
       "3       44  cg03820088     0.337644      7   23522229      TRA2A       TRA2A   \n",
       "4       58  cg04998737     0.336275      5   77645403        OTP         OTP   \n",
       "\n",
       "   ENTREZID_x                              conservationInMouse_x  \\\n",
       "0      440982        mapped to different gene in human and mouse   \n",
       "1         627       conserved gene and region in mouse and human   \n",
       "2      441094        mapped to different gene in human and mouse   \n",
       "3       29896  conserved gene but different region in human a...   \n",
       "4       23440       conserved gene and region in mouse and human   \n",
       "\n",
       "      GeneRegionID_x  ... CHR    bp_hg38  Gene.hg19 ENTREZID  \\\n",
       "0  LOC440982_fiveUTR  ...   3  147409417       ZIC1   440982   \n",
       "1       BDNF_fiveUTR  ...  11   27720483       BDNF      627   \n",
       "2   NR2F1-AS1_Intron  ...   5   93585317      NR2F1   441094   \n",
       "3      TRA2A_fiveUTR  ...   7   23522229      TRA2A    29896   \n",
       "4       OTP_Promoter  ...   5   77645452        OTP    23440   \n",
       "\n",
       "                                 conservationInMouse       GeneRegionID  \\\n",
       "0        mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "1       conserved gene and region in mouse and human       BDNF_fiveUTR   \n",
       "2        mapped to different gene in human and mouse     NR2F1-AS1_Exon   \n",
       "3  conserved gene but different region in human a...      TRA2A_fiveUTR   \n",
       "4       conserved gene and region in mouse and human       OTP_Promoter   \n",
       "\n",
       "         annotation  CHR_mm10      bp_mm10 Gene_mm10  \n",
       "0            5' UTR       9.0   91365746.0      Zic1  \n",
       "1  Promoter (<=1kb)       2.0  109676257.0      Bdnf  \n",
       "2  Promoter (<=1kb)      13.0   78198288.0     Nr2f1  \n",
       "3            5' UTR       6.0   49252905.0     Tra2a  \n",
       "4  Promoter (5-6kb)      13.0   94869014.0       Otp  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Clocks overlap 1, 2 and 3\n",
    "Clock_overlap_1_2_3_top25 = pd.read_csv('../../FunctionalAnalysis/top_25_cpgs_overlap_1_2_3.csv')\n",
    "Clock_overlap_1_2_3_top25.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create lists of genes for all clocks \n",
    "Clock1_genes = Clock1_top25['Gene'].tolist()\n",
    "Clock2_genes = Clock2_top25['Gene'].tolist()\n",
    "Clock3_genes = Clock3_top25['Gene'].tolist()\n",
    "Clock_overlap_2_3_genes = Clock_overlap_2_3_top25['Gene'].tolist()\n",
    "Clock_overlap_1_2_3_genes = Clock_overlap_1_2_3_top25['Gene'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ANK1',\n",
       " 'OTP',\n",
       " 'LOC440982',\n",
       " 'SMUG1',\n",
       " 'DLX1',\n",
       " 'BDNF',\n",
       " 'COL4A1',\n",
       " 'EVX2',\n",
       " 'NR2F1-AS1',\n",
       " 'TRA2A',\n",
       " 'OTP',\n",
       " 'IGF2BP3',\n",
       " 'MGMT',\n",
       " 'LCOR',\n",
       " 'KDM2A',\n",
       " 'ZNF326',\n",
       " 'LHFPL3',\n",
       " 'IQCM',\n",
       " 'FMO3',\n",
       " 'IKZF1',\n",
       " 'CELF6',\n",
       " 'ZIC2',\n",
       " 'CELF6',\n",
       " 'KCNC4',\n",
       " 'RORA']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Clock1_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Call the function to get the data\n",
    "get_geo_data(Clock1_genes)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
