{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis of HUMAN Methylation Data from the Mammalian Methylation Consortium dataset (Horvath 2023)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "need to update the title and description of this notebook as serves mainly to clean and export methylation data (complete, human) and Clock data (complete & human) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import re\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy import stats\n",
    "import statsmodels.stats.multitest as multi\n",
    "from fastcluster import linkage\n",
    "import plotly.graph_objects as go\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Data Cleaning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 Import the datasets (Complete & Human) from the Mammalian Methylation Consortium dataset (Horvath 2023) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the Human CSV file\n",
    "human_methylation_data = pd.read_csv('GSE184211_datBetaNormalized.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the Complete CSV file\n",
    "complete_methylation_data = pd.read_csv('GSE223748_datBetaNormalized.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(37554, 662)\n",
      "(37554, 15044)\n"
     ]
    }
   ],
   "source": [
    "print(human_methylation_data.shape)\n",
    "print(complete_methylation_data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.2. Rename the columns of the datasets to make the datasets more readable and standardized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename column 'CGid' to 'CpG'\n",
    "human_methylation_data.rename(columns={'CGid': 'CpG'}, inplace=True)\n",
    "complete_methylation_data.rename(columns={'Unnamed: 0': 'CpG'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CpG</th>\n",
       "      <th>202897220093_R01C01</th>\n",
       "      <th>202897220093_R02C01</th>\n",
       "      <th>202897220093_R03C01</th>\n",
       "      <th>202897220093_R04C01</th>\n",
       "      <th>202897220093_R05C01</th>\n",
       "      <th>202897220093_R06C01</th>\n",
       "      <th>202897220093_R01C02</th>\n",
       "      <th>202897220093_R02C02</th>\n",
       "      <th>202897220093_R03C02</th>\n",
       "      <th>...</th>\n",
       "      <th>206139140103_R02C02</th>\n",
       "      <th>206139140103_R03C02</th>\n",
       "      <th>206139140103_R04C02</th>\n",
       "      <th>206139140103_R05C02</th>\n",
       "      <th>206139140103_R06C02</th>\n",
       "      <th>206139140104_R01C01</th>\n",
       "      <th>206139140104_R02C01</th>\n",
       "      <th>206139140104_R03C01</th>\n",
       "      <th>206139140104_R04C01</th>\n",
       "      <th>206139140104_R05C01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cg00000165</td>\n",
       "      <td>0.480990</td>\n",
       "      <td>0.483132</td>\n",
       "      <td>0.494042</td>\n",
       "      <td>0.505628</td>\n",
       "      <td>0.496484</td>\n",
       "      <td>0.486760</td>\n",
       "      <td>0.497341</td>\n",
       "      <td>0.482591</td>\n",
       "      <td>0.495475</td>\n",
       "      <td>...</td>\n",
       "      <td>0.140734</td>\n",
       "      <td>0.070563</td>\n",
       "      <td>0.060902</td>\n",
       "      <td>0.056785</td>\n",
       "      <td>0.054896</td>\n",
       "      <td>0.218780</td>\n",
       "      <td>0.141689</td>\n",
       "      <td>0.092939</td>\n",
       "      <td>0.085170</td>\n",
       "      <td>0.060552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cg00001209</td>\n",
       "      <td>0.934681</td>\n",
       "      <td>0.946624</td>\n",
       "      <td>0.940368</td>\n",
       "      <td>0.935963</td>\n",
       "      <td>0.926762</td>\n",
       "      <td>0.931264</td>\n",
       "      <td>0.941756</td>\n",
       "      <td>0.936942</td>\n",
       "      <td>0.940002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.833588</td>\n",
       "      <td>0.872219</td>\n",
       "      <td>0.900707</td>\n",
       "      <td>0.929456</td>\n",
       "      <td>0.861758</td>\n",
       "      <td>0.828209</td>\n",
       "      <td>0.868502</td>\n",
       "      <td>0.825405</td>\n",
       "      <td>0.907443</td>\n",
       "      <td>0.909347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cg00001364</td>\n",
       "      <td>0.911639</td>\n",
       "      <td>0.907298</td>\n",
       "      <td>0.912401</td>\n",
       "      <td>0.901470</td>\n",
       "      <td>0.912350</td>\n",
       "      <td>0.910808</td>\n",
       "      <td>0.911290</td>\n",
       "      <td>0.910110</td>\n",
       "      <td>0.907811</td>\n",
       "      <td>...</td>\n",
       "      <td>0.808296</td>\n",
       "      <td>0.858634</td>\n",
       "      <td>0.840960</td>\n",
       "      <td>0.903902</td>\n",
       "      <td>0.914616</td>\n",
       "      <td>0.778192</td>\n",
       "      <td>0.805107</td>\n",
       "      <td>0.832695</td>\n",
       "      <td>0.889494</td>\n",
       "      <td>0.901459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cg00001582</td>\n",
       "      <td>0.060784</td>\n",
       "      <td>0.057917</td>\n",
       "      <td>0.055478</td>\n",
       "      <td>0.064311</td>\n",
       "      <td>0.060102</td>\n",
       "      <td>0.055498</td>\n",
       "      <td>0.056630</td>\n",
       "      <td>0.060347</td>\n",
       "      <td>0.066302</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077506</td>\n",
       "      <td>0.087474</td>\n",
       "      <td>0.081258</td>\n",
       "      <td>0.067971</td>\n",
       "      <td>0.067423</td>\n",
       "      <td>0.069621</td>\n",
       "      <td>0.075440</td>\n",
       "      <td>0.065853</td>\n",
       "      <td>0.074909</td>\n",
       "      <td>0.074081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cg00002920</td>\n",
       "      <td>0.655778</td>\n",
       "      <td>0.628036</td>\n",
       "      <td>0.697029</td>\n",
       "      <td>0.690823</td>\n",
       "      <td>0.678101</td>\n",
       "      <td>0.615883</td>\n",
       "      <td>0.638421</td>\n",
       "      <td>0.641940</td>\n",
       "      <td>0.654358</td>\n",
       "      <td>...</td>\n",
       "      <td>0.719434</td>\n",
       "      <td>0.782699</td>\n",
       "      <td>0.701002</td>\n",
       "      <td>0.638727</td>\n",
       "      <td>0.755281</td>\n",
       "      <td>0.776376</td>\n",
       "      <td>0.799998</td>\n",
       "      <td>0.781141</td>\n",
       "      <td>0.698296</td>\n",
       "      <td>0.731993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 15044 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CpG  202897220093_R01C01  202897220093_R02C01  202897220093_R03C01  \\\n",
       "0  cg00000165             0.480990             0.483132             0.494042   \n",
       "1  cg00001209             0.934681             0.946624             0.940368   \n",
       "2  cg00001364             0.911639             0.907298             0.912401   \n",
       "3  cg00001582             0.060784             0.057917             0.055478   \n",
       "4  cg00002920             0.655778             0.628036             0.697029   \n",
       "\n",
       "   202897220093_R04C01  202897220093_R05C01  202897220093_R06C01  \\\n",
       "0             0.505628             0.496484             0.486760   \n",
       "1             0.935963             0.926762             0.931264   \n",
       "2             0.901470             0.912350             0.910808   \n",
       "3             0.064311             0.060102             0.055498   \n",
       "4             0.690823             0.678101             0.615883   \n",
       "\n",
       "   202897220093_R01C02  202897220093_R02C02  202897220093_R03C02  ...  \\\n",
       "0             0.497341             0.482591             0.495475  ...   \n",
       "1             0.941756             0.936942             0.940002  ...   \n",
       "2             0.911290             0.910110             0.907811  ...   \n",
       "3             0.056630             0.060347             0.066302  ...   \n",
       "4             0.638421             0.641940             0.654358  ...   \n",
       "\n",
       "   206139140103_R02C02  206139140103_R03C02  206139140103_R04C02  \\\n",
       "0             0.140734             0.070563             0.060902   \n",
       "1             0.833588             0.872219             0.900707   \n",
       "2             0.808296             0.858634             0.840960   \n",
       "3             0.077506             0.087474             0.081258   \n",
       "4             0.719434             0.782699             0.701002   \n",
       "\n",
       "   206139140103_R05C02  206139140103_R06C02  206139140104_R01C01  \\\n",
       "0             0.056785             0.054896             0.218780   \n",
       "1             0.929456             0.861758             0.828209   \n",
       "2             0.903902             0.914616             0.778192   \n",
       "3             0.067971             0.067423             0.069621   \n",
       "4             0.638727             0.755281             0.776376   \n",
       "\n",
       "   206139140104_R02C01  206139140104_R03C01  206139140104_R04C01  \\\n",
       "0             0.141689             0.092939             0.085170   \n",
       "1             0.868502             0.825405             0.907443   \n",
       "2             0.805107             0.832695             0.889494   \n",
       "3             0.075440             0.065853             0.074909   \n",
       "4             0.799998             0.781141             0.698296   \n",
       "\n",
       "   206139140104_R05C01  \n",
       "0             0.060552  \n",
       "1             0.909347  \n",
       "2             0.901459  \n",
       "3             0.074081  \n",
       "4             0.731993  \n",
       "\n",
       "[5 rows x 15044 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_methylation_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.3. Check and handle missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 0 missing values in the complete methylation dataset.\n",
      "There are 0 missing values in the human methylation dataset.\n"
     ]
    }
   ],
   "source": [
    "# Check if there are any missing values\n",
    "complete_missing_values = complete_methylation_data.isnull().sum().sum()\n",
    "human_missing_values = human_methylation_data.isnull().sum().sum()\n",
    "print(f'There are {complete_missing_values} missing values in the complete methylation dataset.')\n",
    "print(f'There are {human_missing_values} missing values in the human methylation dataset.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.4. Check and handle duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 0 duplicated CpG sites in the complete methylation dataset.\n",
      "There are 0 duplicated CpG sites in the human methylation.\n"
     ]
    }
   ],
   "source": [
    "# Check if there are any duplicated CpG sites\n",
    "duplicated_cpg_complete = complete_methylation_data['CpG'].duplicated().sum()\n",
    "duplicated_cpg_human = human_methylation_data['CpG'].duplicated().sum()\n",
    "print(f'There are {duplicated_cpg_complete} duplicated CpG sites in the complete methylation dataset.')\n",
    "print(f'There are {duplicated_cpg_human} duplicated CpG sites in the human methylation.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Exploratory Analysis of the data with Heatmaps and PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.1. Heatmap of the whole dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.2. PCA of the whole dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Filtering the data by the CpGs identified in the Clocks & export clocks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.1. Import CpG's identified by the Clocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import full clocks\n",
    "Clock_1 = pd.read_csv('../FunctionalAnalysis/Clock_1_ordered.csv')\n",
    "Clock_2 = pd.read_csv('../FunctionalAnalysis/Clock_2_ordered.csv')\n",
    "Clock_3 = pd.read_csv('../FunctionalAnalysis/Clock_3_ordered.csv')\n",
    "Clock_1_2_3_overlap = pd.read_csv('../FunctionalAnalysis/overlap_1_2_3_ordered.csv')\n",
    "Clock_2_3_overlap = pd.read_csv('../FunctionalAnalysis/overlap_2_3_ordered.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>143</td>\n",
       "      <td>cg11728741</td>\n",
       "      <td>0.928874</td>\n",
       "      <td>8</td>\n",
       "      <td>41896469</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>286</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>ANK1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>8</td>\n",
       "      <td>22975146.0</td>\n",
       "      <td>Ank1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>297</td>\n",
       "      <td>cg24352905</td>\n",
       "      <td>0.853268</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>160</td>\n",
       "      <td>cg13058338</td>\n",
       "      <td>0.638660</td>\n",
       "      <td>12</td>\n",
       "      <td>54173598</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>23583</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SMUG1_Intron</td>\n",
       "      <td>Intron (ENST00000635234.1/23583, intron 2 of 4)</td>\n",
       "      <td>15</td>\n",
       "      <td>103146996.0</td>\n",
       "      <td>Smug1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>316</td>\n",
       "      <td>cg26067250</td>\n",
       "      <td>0.510613</td>\n",
       "      <td>2</td>\n",
       "      <td>172085721</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>1745</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>DLX1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>71530037.0</td>\n",
       "      <td>Dlx1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock1 CHR    bp_hg38       Gene Gene.hg19  \\\n",
       "0    143  cg11728741     0.928874   8   41896469       ANK1      ANK1   \n",
       "1    297  cg24352905     0.853268   5   77645452        OTP       OTP   \n",
       "2     98  cg08938156     0.843749   3  147409417  LOC440982      ZIC1   \n",
       "3    160  cg13058338     0.638660  12   54173598      SMUG1     SMUG1   \n",
       "4    316  cg26067250     0.510613   2  172085721       DLX1      DLX1   \n",
       "\n",
       "   ENTREZID                           conservationInMouse       GeneRegionID  \\\n",
       "0       286  conserved gene and region in mouse and human          ANK1_Exon   \n",
       "1     23440  conserved gene and region in mouse and human       OTP_Promoter   \n",
       "2    440982   mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "3     23583   mapped to different gene in human and mouse       SMUG1_Intron   \n",
       "4      1745   mapped to different gene in human and mouse          DLX1_Exon   \n",
       "\n",
       "                                        annotation CHR_mm10      bp_mm10  \\\n",
       "0                                 Promoter (<=1kb)        8   22975146.0   \n",
       "1                                 Promoter (5-6kb)       13   94869014.0   \n",
       "2                                           5' UTR        9   91365746.0   \n",
       "3  Intron (ENST00000635234.1/23583, intron 2 of 4)       15  103146996.0   \n",
       "4                                 Promoter (<=1kb)        2   71530037.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0      Ank1  \n",
       "1       Otp  \n",
       "2      Zic1  \n",
       "3     Smug1  \n",
       "4      Dlx1  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Clock_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(335, 14)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Clock_1.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.2. Create subsets of top CpGs identified by clocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a subset of the first 100 and 200 CpG sites for each clock\n",
    "\n",
    "Clock1_100 = Clock_1.head(100)\n",
    "Clock2_100 = Clock_2.head(100)\n",
    "Clock3_100 = Clock_3.head(100)\n",
    "Clock1_2_3_overlap_100 = Clock_1_2_3_overlap.head(100)\n",
    "Clock2_3_overlap_100 = Clock_2_3_overlap.head(100)\n",
    "\n",
    "Clock1_200 = Clock_1.head(200)\n",
    "Clock2_200 = Clock_2.head(200)\n",
    "Clock3_200 = Clock_3.head(200)\n",
    "Clock1_2_3_overlap_200 = Clock_1_2_3_overlap.head(200)\n",
    "Clock2_3_overlap_200 = Clock_2_3_overlap.head(200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(100, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>143</td>\n",
       "      <td>cg11728741</td>\n",
       "      <td>0.928874</td>\n",
       "      <td>8</td>\n",
       "      <td>41896469</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>286</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>ANK1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>8</td>\n",
       "      <td>22975146.0</td>\n",
       "      <td>Ank1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>297</td>\n",
       "      <td>cg24352905</td>\n",
       "      <td>0.853268</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>160</td>\n",
       "      <td>cg13058338</td>\n",
       "      <td>0.638660</td>\n",
       "      <td>12</td>\n",
       "      <td>54173598</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>23583</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SMUG1_Intron</td>\n",
       "      <td>Intron (ENST00000635234.1/23583, intron 2 of 4)</td>\n",
       "      <td>15</td>\n",
       "      <td>103146996.0</td>\n",
       "      <td>Smug1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>316</td>\n",
       "      <td>cg26067250</td>\n",
       "      <td>0.510613</td>\n",
       "      <td>2</td>\n",
       "      <td>172085721</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>1745</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>DLX1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>71530037.0</td>\n",
       "      <td>Dlx1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock1 CHR    bp_hg38       Gene Gene.hg19  \\\n",
       "0    143  cg11728741     0.928874   8   41896469       ANK1      ANK1   \n",
       "1    297  cg24352905     0.853268   5   77645452        OTP       OTP   \n",
       "2     98  cg08938156     0.843749   3  147409417  LOC440982      ZIC1   \n",
       "3    160  cg13058338     0.638660  12   54173598      SMUG1     SMUG1   \n",
       "4    316  cg26067250     0.510613   2  172085721       DLX1      DLX1   \n",
       "\n",
       "   ENTREZID                           conservationInMouse       GeneRegionID  \\\n",
       "0       286  conserved gene and region in mouse and human          ANK1_Exon   \n",
       "1     23440  conserved gene and region in mouse and human       OTP_Promoter   \n",
       "2    440982   mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "3     23583   mapped to different gene in human and mouse       SMUG1_Intron   \n",
       "4      1745   mapped to different gene in human and mouse          DLX1_Exon   \n",
       "\n",
       "                                        annotation CHR_mm10      bp_mm10  \\\n",
       "0                                 Promoter (<=1kb)        8   22975146.0   \n",
       "1                                 Promoter (5-6kb)       13   94869014.0   \n",
       "2                                           5' UTR        9   91365746.0   \n",
       "3  Intron (ENST00000635234.1/23583, intron 2 of 4)       15  103146996.0   \n",
       "4                                 Promoter (<=1kb)        2   71530037.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0      Ank1  \n",
       "1       Otp  \n",
       "2      Zic1  \n",
       "3     Smug1  \n",
       "4      Dlx1  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sneak peek of the top CpGs for clock 1\n",
    "print(Clock1_100.shape)\n",
    "Clock1_100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(100, 40)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>329</td>\n",
       "      <td>cg27201382</td>\n",
       "      <td>0.444645</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>109676257.0</td>\n",
       "      <td>Bdnf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>144</td>\n",
       "      <td>cg11904056</td>\n",
       "      <td>0.341791</td>\n",
       "      <td>5</td>\n",
       "      <td>93583204</td>\n",
       "      <td>NR2F1-AS1</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>93585317</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>78198288.0</td>\n",
       "      <td>Nr2f1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>44</td>\n",
       "      <td>cg03820088</td>\n",
       "      <td>0.337644</td>\n",
       "      <td>7</td>\n",
       "      <td>23522229</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>29896</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>TRA2A_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>23522229</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>29896</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>TRA2A_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>6</td>\n",
       "      <td>49252905.0</td>\n",
       "      <td>Tra2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>58</td>\n",
       "      <td>cg04998737</td>\n",
       "      <td>0.336275</td>\n",
       "      <td>5</td>\n",
       "      <td>77645403</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_x       var_x  beta_clock1 CHR_x  bp_hg38_x       Gene Gene.hg19_x  \\\n",
       "0       98  cg08938156     0.843749     3  147409417  LOC440982        ZIC1   \n",
       "1      329  cg27201382     0.444645    11   27720483       BDNF        BDNF   \n",
       "2      144  cg11904056     0.341791     5   93583204  NR2F1-AS1       NR2F1   \n",
       "3       44  cg03820088     0.337644     7   23522229      TRA2A       TRA2A   \n",
       "4       58  cg04998737     0.336275     5   77645403        OTP         OTP   \n",
       "\n",
       "   ENTREZID_x                              conservationInMouse_x  \\\n",
       "0      440982        mapped to different gene in human and mouse   \n",
       "1         627       conserved gene and region in mouse and human   \n",
       "2      441094        mapped to different gene in human and mouse   \n",
       "3       29896  conserved gene but different region in human a...   \n",
       "4       23440       conserved gene and region in mouse and human   \n",
       "\n",
       "      GeneRegionID_x  ... CHR    bp_hg38  Gene.hg19 ENTREZID  \\\n",
       "0  LOC440982_fiveUTR  ...   3  147409417       ZIC1   440982   \n",
       "1       BDNF_fiveUTR  ...  11   27720483       BDNF      627   \n",
       "2   NR2F1-AS1_Intron  ...   5   93585317      NR2F1   441094   \n",
       "3      TRA2A_fiveUTR  ...   7   23522229      TRA2A    29896   \n",
       "4       OTP_Promoter  ...   5   77645452        OTP    23440   \n",
       "\n",
       "                                 conservationInMouse       GeneRegionID  \\\n",
       "0        mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "1       conserved gene and region in mouse and human       BDNF_fiveUTR   \n",
       "2        mapped to different gene in human and mouse     NR2F1-AS1_Exon   \n",
       "3  conserved gene but different region in human a...      TRA2A_fiveUTR   \n",
       "4       conserved gene and region in mouse and human       OTP_Promoter   \n",
       "\n",
       "         annotation CHR_mm10      bp_mm10 Gene_mm10  \n",
       "0            5' UTR        9   91365746.0      Zic1  \n",
       "1  Promoter (<=1kb)        2  109676257.0      Bdnf  \n",
       "2  Promoter (<=1kb)       13   78198288.0     Nr2f1  \n",
       "3            5' UTR        6   49252905.0     Tra2a  \n",
       "4  Promoter (5-6kb)       13   94869014.0       Otp  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sneak peek of the overlaps (CpG col has a different name)\n",
    "print(Clock1_2_3_overlap_100.shape)\n",
    "Clock1_2_3_overlap_100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['index_x', 'var_x', 'beta_clock1', 'CHR_x', 'bp_hg38_x', 'Gene',\n",
      "       'Gene.hg19_x', 'ENTREZID_x', 'conservationInMouse_x', 'GeneRegionID_x',\n",
      "       'annotation_x', 'CHR_mm10_x', 'bp_mm10_x', 'Gene_mm10_x', 'index_y',\n",
      "       'var_y', 'beta_clock2', 'CHR_y', 'bp_hg38_y', 'Gene.hg19_y',\n",
      "       'ENTREZID_y', 'conservationInMouse_y', 'GeneRegionID_y', 'annotation_y',\n",
      "       'CHR_mm10_y', 'bp_mm10_y', 'Gene_mm10_y', 'index', 'var', 'beta_clock3',\n",
      "       'CHR', 'bp_hg38', 'Gene.hg19', 'ENTREZID', 'conservationInMouse',\n",
      "       'GeneRegionID', 'annotation', 'CHR_mm10', 'bp_mm10', 'Gene_mm10'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Print colnames\n",
    "print(Clock1_2_3_overlap_100.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.3. Create lists with the CpGs identified by the clocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a list of the CpGs (var) for each clock \n",
    "\n",
    "Clock1_CpG_list = Clock_1['var'].tolist()\n",
    "Clock2_CpG_list = Clock_2['var'].tolist()\n",
    "Clock3_CpG_list = Clock_3['var'].tolist()\n",
    "Clock1_2_3_overlap_CpG_list = Clock_1_2_3_overlap['var_x'].tolist()\n",
    "Clock2_3_overlap_CpG_list = Clock_2_3_overlap['var_x'].tolist()\n",
    "\n",
    "Clock1_100_CpG_list = Clock1_100['var'].tolist()\n",
    "Clock2_100_CpG_list = Clock2_100['var'].tolist()\n",
    "Clock3_100_CpG_list = Clock3_100['var'].tolist()\n",
    "Clock1_2_3_overlap_100_CpG_list = Clock1_2_3_overlap_100['var_x'].tolist()\n",
    "Clock2_3_overlap_100_CpG_list = Clock2_3_overlap_100['var_x'].tolist()\n",
    "\n",
    "Clock1_200_CpG_list = Clock1_200['var'].tolist()\n",
    "Clock2_200_CpG_list = Clock2_200['var'].tolist()\n",
    "Clock3_200_CpG_list = Clock3_200['var'].tolist()\n",
    "Clock1_2_3_overlap_200_CpG_list = Clock1_2_3_overlap_200['var_x'].tolist()\n",
    "Clock2_3_overlap_200_CpG_list = Clock2_3_overlap_200['var_x'].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['cg11728741', 'cg24352905', 'cg08938156', 'cg13058338', 'cg26067250', 'cg27201382', 'cg00593462', 'cg09227056', 'cg11904056', 'cg03820088', 'cg04998737', 'cg08074329', 'cg17143801', 'cg04368876', 'cg05401971', 'cg16960327', 'cg09710440', 'cg03605454', 'cg16599143', 'cg22661206', 'cg18821963', 'cg15682828', 'cg23087015', 'cg12651099', 'cg12156848', 'cg09622673', 'cg22264409', 'cg11084334', 'cg15340018', 'cg23985931', 'cg08676748', 'cg03370924', 'cg01429475', 'cg12373771', 'cg25943666', 'cg00249943', 'cg14381350', 'cg26844246', 'cg23076498', 'cg16867657', 'cg25019875', 'cg09497746', 'cg04097289', 'cg02532525', 'cg17267107', 'cg05475048', 'cg10125423', 'cg12841266', 'cg15437942', 'cg25468783', 'cg12830057', 'cg04211240', 'cg15719056', 'cg01009870', 'cg18194685', 'cg26012482', 'cg05575054', 'cg27553955', 'cg03564272', 'cg00513357', 'cg24357465', 'cg19981759', 'cg26512254', 'cg03942000', 'cg03185524', 'cg14384416', 'cg24952195', 'cg07499079', 'cg10040493', 'cg25217825', 'cg16570119', 'cg18304538', 'cg10160701', 'cg01153166', 'cg22031066', 'cg08216101', 'cg21345677', 'cg24504442', 'cg12815291', 'cg17116410', 'cg17114460', 'cg26311454', 'cg24655241', 'cg11666770', 'cg01698540', 'cg23415880', 'cg11235379', 'cg20738648', 'cg17718834', 'cg05562063', 'cg23597501', 'cg02474352', 'cg00587168', 'cg01975510', 'cg09568652', 'cg01246498', 'cg18094982', 'cg12449245', 'cg22457719', 'cg00362836']\n"
     ]
    }
   ],
   "source": [
    "print(Clock1_100_CpG_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.4. Filter the full and human methylation datasets by the CpG lists & save resulting datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering the full methylation data by the FULL Clocks\n",
    "\n",
    "Clock1_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock1_CpG_list)]\n",
    "Clock1_methylation_data_complete = Clock1_methylation_data_complete.set_index('CpG').loc[Clock1_CpG_list].reset_index()\n",
    "Clock1_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock1_CpG_list)]\n",
    "Clock1_methylation_data_human = Clock1_methylation_data_human.set_index('CpG').loc[Clock1_CpG_list].reset_index()\n",
    "\n",
    "Clock2_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock2_CpG_list)]\n",
    "Clock2_methylation_data_complete = Clock2_methylation_data_complete.set_index('CpG').loc[Clock2_CpG_list].reset_index()\n",
    "Clock2_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock2_CpG_list)]\n",
    "Clock2_methylation_data_human = Clock2_methylation_data_human.set_index('CpG').loc[Clock2_CpG_list].reset_index()\n",
    "\n",
    "Clock3_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock3_CpG_list)]\n",
    "Clock3_methylation_data_complete = Clock3_methylation_data_complete.set_index('CpG').loc[Clock3_CpG_list].reset_index()\n",
    "Clock3_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock3_CpG_list)]\n",
    "Clock3_methylation_data_human = Clock3_methylation_data_human.set_index('CpG').loc[Clock3_CpG_list].reset_index()\n",
    "\n",
    "Clock1_2_3_overlap_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock1_2_3_overlap_CpG_list)]\n",
    "Clock1_2_3_overlap_methylation_data_complete = Clock1_2_3_overlap_methylation_data_complete.set_index('CpG').loc[Clock1_2_3_overlap_CpG_list].reset_index()\n",
    "Clock1_2_3_overlap_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock1_2_3_overlap_CpG_list)]\n",
    "Clock1_2_3_overlap_methylation_data_human = Clock1_2_3_overlap_methylation_data_human.set_index('CpG').loc[Clock1_2_3_overlap_CpG_list].reset_index()\n",
    "\n",
    "Clock2_3_overlap_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock2_3_overlap_CpG_list)]\n",
    "Clock2_3_overlap_methylation_data_complete = Clock2_3_overlap_methylation_data_complete.set_index('CpG').loc[Clock2_3_overlap_CpG_list].reset_index()\n",
    "Clock2_3_overlap_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock2_3_overlap_CpG_list)]\n",
    "Clock2_3_overlap_methylation_data_human = Clock2_3_overlap_methylation_data_human.set_index('CpG').loc[Clock2_3_overlap_CpG_list].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Filtering the full methylation data by the Clocks top 100 CpG's\n",
    "\n",
    "Clock1_100_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock1_100_CpG_list)]\n",
    "Clock1_100_methylation_data_complete = Clock1_100_methylation_data_complete.set_index('CpG').loc[Clock1_100_CpG_list].reset_index()\n",
    "Clock1_100_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock1_100_CpG_list)]\n",
    "Clock1_100_methylation_data_human = Clock1_100_methylation_data_human.set_index('CpG').loc[Clock1_100_CpG_list].reset_index()\n",
    "\n",
    "Clock2_100_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock2_100_CpG_list)]\n",
    "Clock2_100_methylation_data_complete = Clock2_100_methylation_data_complete.set_index('CpG').loc[Clock2_100_CpG_list].reset_index()\n",
    "Clock2_100_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock2_100_CpG_list)]\n",
    "Clock2_100_methylation_data_human = Clock2_100_methylation_data_human.set_index('CpG').loc[Clock2_100_CpG_list].reset_index()\n",
    "\n",
    "Clock3_100_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock3_100_CpG_list)]\n",
    "Clock3_100_methylation_data_complete = Clock3_100_methylation_data_complete.set_index('CpG').loc[Clock3_100_CpG_list].reset_index()\n",
    "Clock3_100_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock3_100_CpG_list)]\n",
    "Clock3_100_methylation_data_human = Clock3_100_methylation_data_human.set_index('CpG').loc[Clock3_100_CpG_list].reset_index()\n",
    "\n",
    "Clock1_2_3_overlap_100_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock1_2_3_overlap_100_CpG_list)]\n",
    "Clock1_2_3_overlap_100_methylation_data_complete = Clock1_2_3_overlap_100_methylation_data_complete.set_index('CpG').loc[Clock1_2_3_overlap_100_CpG_list].reset_index()\n",
    "Clock1_2_3_overlap_100_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock1_2_3_overlap_100_CpG_list)]\n",
    "Clock1_2_3_overlap_100_methylation_data_human = Clock1_2_3_overlap_100_methylation_data_human.set_index('CpG').loc[Clock1_2_3_overlap_100_CpG_list].reset_index()\n",
    "\n",
    "Clock2_3_overlap_100_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock2_3_overlap_100_CpG_list)]\n",
    "Clock2_3_overlap_100_methylation_data_complete = Clock2_3_overlap_100_methylation_data_complete.set_index('CpG').loc[Clock2_3_overlap_100_CpG_list].reset_index()\n",
    "Clock2_3_overlap_100_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock2_3_overlap_100_CpG_list)]\n",
    "Clock2_3_overlap_100_methylation_data_human = Clock2_3_overlap_100_methylation_data_human.set_index('CpG').loc[Clock2_3_overlap_100_CpG_list].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering the full methylation data by the Clock's top 200 CpGs\n",
    "\n",
    "Clock1_200_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock1_200_CpG_list)]\n",
    "Clock1_200_methylation_data_complete = Clock1_200_methylation_data_complete.set_index('CpG').loc[Clock1_200_CpG_list].reset_index()\n",
    "Clock1_200_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock1_200_CpG_list)]\n",
    "Clock1_200_methylation_data_human = Clock1_200_methylation_data_human.set_index('CpG').loc[Clock1_200_CpG_list].reset_index()\n",
    "\n",
    "Clock2_200_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock2_200_CpG_list)]\n",
    "Clock2_200_methylation_data_complete = Clock2_200_methylation_data_complete.set_index('CpG').loc[Clock2_200_CpG_list].reset_index()\n",
    "Clock2_200_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock2_200_CpG_list)]\n",
    "Clock2_200_methylation_data_human = Clock2_200_methylation_data_human.set_index('CpG').loc[Clock2_200_CpG_list].reset_index()\n",
    "\n",
    "Clock3_200_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock3_200_CpG_list)]\n",
    "Clock3_200_methylation_data_complete = Clock3_200_methylation_data_complete.set_index('CpG').loc[Clock3_200_CpG_list].reset_index()\n",
    "Clock3_200_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock3_200_CpG_list)]\n",
    "Clock3_200_methylation_data_human = Clock3_200_methylation_data_human.set_index('CpG').loc[Clock3_200_CpG_list].reset_index()\n",
    "\n",
    "Clock1_2_3_overlap_200_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock1_2_3_overlap_200_CpG_list)]\n",
    "Clock1_2_3_overlap_200_methylation_data_complete = Clock1_2_3_overlap_200_methylation_data_complete.set_index('CpG').loc[Clock1_2_3_overlap_200_CpG_list].reset_index()\n",
    "Clock1_2_3_overlap_200_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock1_2_3_overlap_200_CpG_list)]\n",
    "Clock1_2_3_overlap_200_methylation_data_human = Clock1_2_3_overlap_200_methylation_data_human.set_index('CpG').loc[Clock1_2_3_overlap_200_CpG_list].reset_index()\n",
    "\n",
    "Clock2_3_overlap_200_methylation_data_complete = complete_methylation_data[complete_methylation_data['CpG'].isin(Clock2_3_overlap_200_CpG_list)]\n",
    "Clock2_3_overlap_200_methylation_data_complete = Clock2_3_overlap_200_methylation_data_complete.set_index('CpG').loc[Clock2_3_overlap_200_CpG_list].reset_index()\n",
    "Clock2_3_overlap_200_methylation_data_human = human_methylation_data[human_methylation_data['CpG'].isin(Clock2_3_overlap_200_CpG_list)] \n",
    "Clock2_3_overlap_200_methylation_data_human = Clock2_3_overlap_200_methylation_data_human.set_index('CpG').loc[Clock2_3_overlap_200_CpG_list].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the data - Complete Clocks\n",
    "Clock1_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock1_methylation_data_complete.csv', index=False)\n",
    "Clock1_methylation_data_human.to_csv('./Clocks Methylation Data/Clock1_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock2_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock2_methylation_data_complete.csv', index=False)\n",
    "Clock2_methylation_data_human.to_csv('./Clocks Methylation Data/Clock2_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock3_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock3_methylation_data_complete.csv', index=False)\n",
    "Clock3_methylation_data_human.to_csv('./Clocks Methylation Data/Clock3_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock1_2_3_overlap_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock1_2_3_overlap_methylation_data_complete.csv', index=False)\n",
    "Clock1_2_3_overlap_methylation_data_human.to_csv('./Clocks Methylation Data/Clock1_2_3_overlap_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock2_3_overlap_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock2_3_overlap_methylation_data_complete.csv', index=False)\n",
    "Clock2_3_overlap_methylation_data_human.to_csv('./Clocks Methylation Data/Clock2_3_overlap_methylation_data_human.csv', index=False)       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the data - Top 100 CpGs\n",
    "Clock1_100_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock1_100_methylation_data_complete.csv', index=False)\n",
    "Clock1_100_methylation_data_human.to_csv('./Clocks Methylation Data/Clock1_100_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock2_100_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock2_100_methylation_data_complete.csv', index=False)\n",
    "Clock2_100_methylation_data_human.to_csv('./Clocks Methylation Data/Clock2_100_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock3_100_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock3_100_methylation_data_complete.csv', index=False)\n",
    "Clock3_100_methylation_data_human.to_csv('./Clocks Methylation Data/Clock3_100_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock1_2_3_overlap_100_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock1_2_3_overlap_100_methylation_data_complete.csv', index=False)\n",
    "Clock1_2_3_overlap_100_methylation_data_human.to_csv('./Clocks Methylation Data/Clock1_2_3_overlap_100_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock2_3_overlap_100_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock2_3_overlap_100_methylation_data_complete.csv', index=False)\n",
    "Clock2_3_overlap_100_methylation_data_human.to_csv('./Clocks Methylation Data/Clock2_3_overlap_100_methylation_data_human.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the data - Top 200 CpGs\n",
    "Clock1_200_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock1_200_methylation_data_complete.csv', index=False)\n",
    "Clock1_200_methylation_data_human.to_csv('./Clocks Methylation Data/Clock1_200_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock2_200_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock2_200_methylation_data_complete.csv', index=False)\n",
    "Clock2_200_methylation_data_human.to_csv('./Clocks Methylation Data/Clock2_200_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock3_200_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock3_200_methylation_data_complete.csv', index=False)\n",
    "Clock3_200_methylation_data_human.to_csv('./Clocks Methylation Data/Clock3_200_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock1_2_3_overlap_200_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock1_2_3_overlap_200_methylation_data_complete.csv', index=False)\n",
    "Clock1_2_3_overlap_200_methylation_data_human.to_csv('./Clocks Methylation Data/Clock1_2_3_overlap_200_methylation_data_human.csv', index=False)\n",
    "\n",
    "Clock2_3_overlap_200_methylation_data_complete.to_csv('./Clocks Methylation Data/Clock2_3_overlap_200_methylation_data_complete.csv', index=False)\n",
    "Clock2_3_overlap_200_methylation_data_human.to_csv('./Clocks Methylation Data/Clock2_3_overlap_200_methylation_data_human.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Adding metadata information about each sample of the Human dataset \n",
    "(Note: there is a more up to data metadata file with the complete mammal metadata, including human)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1. Scrapping metadata from GEO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making a list with the sample names (without the first column - CpG names)   \n",
    "sample_names = methylation_data.columns[1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 661\n",
      "Samples found: 661\n",
      "Samples not found: 0\n",
      "List of Samples not found: []\n"
     ]
    }
   ],
   "source": [
    "# Scraping metadata for all human samples in the dataset\n",
    "\n",
    "metadata_human_samples = []\n",
    "\n",
    "for page_num in range(1, 3):\n",
    "    try:\n",
    "        url = f'https://www.ncbi.nlm.nih.gov/geo/browse/?view=samples&display=1000&series=184211&page={page_num}'\n",
    "        response = requests.get(url)\n",
    "        response.raise_for_status()\n",
    "        soup = BeautifulSoup(response.content, 'html.parser')\n",
    "        html_text = soup.get_text()\n",
    "\n",
    "        for sample_name in sample_names:\n",
    "            search_term = sample_name.split('X')[1]\n",
    "            pattern = r'\\b(GSM\\d+)\\s+Sample\\d+\\.Human\\.{}\\b'.format(re.escape(search_term))\n",
    "            match = re.search(pattern, html_text)\n",
    "            \n",
    "            if match:\n",
    "                geo_id = match.group(1)\n",
    "                url = f'https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc={geo_id}'\n",
    "                response = requests.get(url)\n",
    "                response.raise_for_status()\n",
    "                soup = BeautifulSoup(response.content, 'html.parser')\n",
    "                \n",
    "                description_tag = soup.find('td', string=lambda text: text and 'Description' in text)\n",
    "                description = description_tag.find_next('td').text.strip()\n",
    "                characteristics_tag = soup.find('td', string=lambda text: text and 'Characteristics' in text)\n",
    "                characteristics = characteristics_tag.find_next('td').text.strip()\n",
    "                \n",
    "                tissue = description.split('tissue=')[1].split(' ')[0]\n",
    "                age = characteristics.split('age: ')[1].split('confidenceinageestimate')[0].strip()\n",
    "                sex = characteristics.split('Sex: ')[1].split('\\n')[0]\n",
    "                \n",
    "                metadata_human_samples.append([sample_name, geo_id, tissue, age, sex])\n",
    "    \n",
    "    except Exception as e:\n",
    "        print(f\"Error processing page {page_num}: {e}\")\n",
    "\n",
    "# Check if metadata list is empty\n",
    "if not metadata_human_samples:\n",
    "    print(\"No samples found.\")\n",
    "else:\n",
    "    # Create a pandas DataFrame from the metadata list\n",
    "    human_metadata_df = pd.DataFrame(metadata_human_samples, columns=['Sample', 'GEO_ID', 'Tissue', 'Age', 'Sex'])\n",
    "    # Print the DataFrame\n",
    "    # print(metadata_df)\n",
    "\n",
    "# Check for missing samples\n",
    "found_samples = human_metadata_df['Sample'].tolist() if not human_metadata_df.empty else []\n",
    "missing_samples = [sample_name for sample_name in sample_names if sample_name not in found_samples]\n",
    "\n",
    "# Print counts of samples found and not found\n",
    "print(\"Total samples:\", len(sample_names))\n",
    "print(\"Samples found:\", len(found_samples))\n",
    "print(\"Samples not found:\", len(missing_samples))\n",
    "print('List of Samples not found:', missing_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>GEO_ID</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X203203210100_R01C01</td>\n",
       "      <td>GSM5580349</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>50</td>\n",
       "      <td>Male</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X203203210100_R02C01</td>\n",
       "      <td>GSM5580350</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>44</td>\n",
       "      <td>Female</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X203203210100_R03C01</td>\n",
       "      <td>GSM5580351</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>52</td>\n",
       "      <td>Female</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X203203210100_R04C01</td>\n",
       "      <td>GSM5580352</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>62</td>\n",
       "      <td>Female</td>\n",
       "      <td>60-90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X203203210100_R05C01</td>\n",
       "      <td>GSM5580353</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>65</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Sample      GEO_ID   Tissue  Age     Sex Age Group\n",
       "0  X203203210100_R01C01  GSM5580349  Adipose   50    Male     30-60\n",
       "1  X203203210100_R02C01  GSM5580350  Adipose   44  Female     30-60\n",
       "2  X203203210100_R03C01  GSM5580351  Adipose   52  Female     30-60\n",
       "3  X203203210100_R04C01  GSM5580352  Adipose   62  Female     60-90\n",
       "4  X203203210100_R05C01  GSM5580353  Adipose   65    Male     60-90"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sneak peek of the metadata\n",
    "human_metadata_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(661, 6)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Shape of the human metadata dataframe\n",
    "human_metadata_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the number of unique tissues present in the metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique tissue types in the metadata: 11\n"
     ]
    }
   ],
   "source": [
    "# Check the number of unique tissues\n",
    "unique_tissues = human_metadata_df['Tissue'].nunique()\n",
    "print(f'Unique tissue types in the metadata: {unique_tissues}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "Lung          113\n",
       "Heart          97\n",
       "Kidney         97\n",
       "Liver          97\n",
       "Spleen         62\n",
       "Adipose        57\n",
       "Muscle         57\n",
       "Blood          32\n",
       "LymphNode      27\n",
       "BoneMarrow     20\n",
       "Pituitary       2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many samples are there for each tissue type?\n",
    "tissue_counts = human_metadata_df['Tissue'].value_counts()\n",
    "tissue_counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the range of ages and decide if I can split the ages into groups (possibly less than 30 and more than 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimum age: 26\n",
      "Maximum age: 91\n"
     ]
    }
   ],
   "source": [
    "# Check the min ad max age\n",
    "min_age = human_metadata_df['Age'].min()\n",
    "max_age = human_metadata_df['Age'].max()\n",
    "print(f'Minimum age: {min_age}')\n",
    "print(f'Maximum age: {max_age}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples for which ages is less than 30: 13\n",
      "Proportion of samples with age less than 30: 0.019667170953101363\n"
     ]
    }
   ],
   "source": [
    "# Proportion of samples for which ages is less than 30\n",
    "age_less_than_30 = human_metadata_df\n",
    "age_less_than_30['Age'] = pd.to_numeric(age_less_than_30['Age'], errors='coerce')\n",
    "age_less_than_30 = age_less_than_30[age_less_than_30['Age'] < 30]\n",
    "proportion_age_less_than_30 = len(age_less_than_30) / len(human_metadata_df)\n",
    "print(f'Number of samples for which ages is less than 30: {len(age_less_than_30)}')\n",
    "print(f'Proportion of samples with age less than 30: {proportion_age_less_than_30}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Age Group\n",
      "0-30      13\n",
      "30-60    356\n",
      "60-90    289\n",
      "90+        3\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Separate the data into age groups (0-30, 30-60, 60-90, 90+)\n",
    "age_groups = []\n",
    "\n",
    "for age in human_metadata_df['Age']:\n",
    "    if age < 30:\n",
    "        age_group = '0-30'\n",
    "    elif age < 60:\n",
    "        age_group = '30-60'\n",
    "    elif age < 90:\n",
    "        age_group = '60-90'\n",
    "    else:\n",
    "        age_group = '90+'\n",
    "    age_groups.append(age_group)\n",
    "\n",
    "# Add the age groups to the metadata DataFrame\n",
    "human_metadata_df['Age Group'] = age_groups\n",
    "\n",
    "# Check the number of samples in each age group\n",
    "age_group_counts = human_metadata_df\n",
    "age_group_counts = age_group_counts.groupby('Age Group').size()\n",
    "print(age_group_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>GEO_ID</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X203203210100_R01C01</td>\n",
       "      <td>GSM5580349</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>50</td>\n",
       "      <td>Male</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X203203210100_R02C01</td>\n",
       "      <td>GSM5580350</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>44</td>\n",
       "      <td>Female</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X203203210100_R03C01</td>\n",
       "      <td>GSM5580351</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>52</td>\n",
       "      <td>Female</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X203203210100_R04C01</td>\n",
       "      <td>GSM5580352</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>62</td>\n",
       "      <td>Female</td>\n",
       "      <td>60-90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X203203210100_R05C01</td>\n",
       "      <td>GSM5580353</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>65</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Sample      GEO_ID   Tissue  Age     Sex Age Group\n",
       "0  X203203210100_R01C01  GSM5580349  Adipose   50    Male     30-60\n",
       "1  X203203210100_R02C01  GSM5580350  Adipose   44  Female     30-60\n",
       "2  X203203210100_R03C01  GSM5580351  Adipose   52  Female     30-60\n",
       "3  X203203210100_R04C01  GSM5580352  Adipose   62  Female     60-90\n",
       "4  X203203210100_R05C01  GSM5580353  Adipose   65    Male     60-90"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add age group to the metadata\n",
    "human_metadata_df['Age Group'] = age_groups\n",
    "human_metadata_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the metadata to a CSV file\n",
    "human_metadata_df.to_csv('human_metadata_184211.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import back if needed (so we don't have to scrape again)\n",
    "human_metadata_df = pd.read_csv('human_metadata_184211.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CpG</th>\n",
       "      <th>X203203210100_R01C01</th>\n",
       "      <th>X203203210100_R02C01</th>\n",
       "      <th>X203203210100_R03C01</th>\n",
       "      <th>X203203210100_R04C01</th>\n",
       "      <th>X203203210100_R05C01</th>\n",
       "      <th>X203203210100_R06C01</th>\n",
       "      <th>X203203210100_R01C02</th>\n",
       "      <th>X203203210100_R02C02</th>\n",
       "      <th>X203203210100_R03C02</th>\n",
       "      <th>...</th>\n",
       "      <th>X203203210013_R03C01</th>\n",
       "      <th>X203203210013_R04C01</th>\n",
       "      <th>X203203210013_R05C01</th>\n",
       "      <th>X203203210013_R06C01</th>\n",
       "      <th>X203203210013_R01C02</th>\n",
       "      <th>X203203210013_R02C02</th>\n",
       "      <th>X203203210013_R03C02</th>\n",
       "      <th>X203203210013_R04C02</th>\n",
       "      <th>X203203210013_R05C02</th>\n",
       "      <th>X203203210013_R06C02</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cg00000165</td>\n",
       "      <td>0.101505</td>\n",
       "      <td>0.109506</td>\n",
       "      <td>0.141691</td>\n",
       "      <td>0.099335</td>\n",
       "      <td>0.119663</td>\n",
       "      <td>0.133722</td>\n",
       "      <td>0.160619</td>\n",
       "      <td>0.103879</td>\n",
       "      <td>0.113632</td>\n",
       "      <td>...</td>\n",
       "      <td>0.134059</td>\n",
       "      <td>0.111548</td>\n",
       "      <td>0.127344</td>\n",
       "      <td>0.084265</td>\n",
       "      <td>0.112711</td>\n",
       "      <td>0.098072</td>\n",
       "      <td>0.096010</td>\n",
       "      <td>0.117737</td>\n",
       "      <td>0.093164</td>\n",
       "      <td>0.184909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cg00001209</td>\n",
       "      <td>0.874657</td>\n",
       "      <td>0.899091</td>\n",
       "      <td>0.856218</td>\n",
       "      <td>0.896215</td>\n",
       "      <td>0.881654</td>\n",
       "      <td>0.919595</td>\n",
       "      <td>0.897742</td>\n",
       "      <td>0.895401</td>\n",
       "      <td>0.888192</td>\n",
       "      <td>...</td>\n",
       "      <td>0.857043</td>\n",
       "      <td>0.826831</td>\n",
       "      <td>0.866019</td>\n",
       "      <td>0.831738</td>\n",
       "      <td>0.852618</td>\n",
       "      <td>0.906398</td>\n",
       "      <td>0.884948</td>\n",
       "      <td>0.846944</td>\n",
       "      <td>0.924727</td>\n",
       "      <td>0.873856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cg00001364</td>\n",
       "      <td>0.809139</td>\n",
       "      <td>0.841104</td>\n",
       "      <td>0.768823</td>\n",
       "      <td>0.812510</td>\n",
       "      <td>0.767511</td>\n",
       "      <td>0.855345</td>\n",
       "      <td>0.825716</td>\n",
       "      <td>0.876175</td>\n",
       "      <td>0.845755</td>\n",
       "      <td>...</td>\n",
       "      <td>0.848691</td>\n",
       "      <td>0.865939</td>\n",
       "      <td>0.903710</td>\n",
       "      <td>0.920990</td>\n",
       "      <td>0.860893</td>\n",
       "      <td>0.895960</td>\n",
       "      <td>0.894021</td>\n",
       "      <td>0.896620</td>\n",
       "      <td>0.906478</td>\n",
       "      <td>0.891837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cg00001582</td>\n",
       "      <td>0.040719</td>\n",
       "      <td>0.042991</td>\n",
       "      <td>0.041190</td>\n",
       "      <td>0.047361</td>\n",
       "      <td>0.043020</td>\n",
       "      <td>0.043740</td>\n",
       "      <td>0.044908</td>\n",
       "      <td>0.037882</td>\n",
       "      <td>0.046859</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035247</td>\n",
       "      <td>0.034831</td>\n",
       "      <td>0.039376</td>\n",
       "      <td>0.040148</td>\n",
       "      <td>0.040499</td>\n",
       "      <td>0.039100</td>\n",
       "      <td>0.040631</td>\n",
       "      <td>0.043228</td>\n",
       "      <td>0.045163</td>\n",
       "      <td>0.040598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cg00002920</td>\n",
       "      <td>0.415627</td>\n",
       "      <td>0.562619</td>\n",
       "      <td>0.492849</td>\n",
       "      <td>0.537996</td>\n",
       "      <td>0.646229</td>\n",
       "      <td>0.324161</td>\n",
       "      <td>0.240706</td>\n",
       "      <td>0.492182</td>\n",
       "      <td>0.404285</td>\n",
       "      <td>...</td>\n",
       "      <td>0.329255</td>\n",
       "      <td>0.315051</td>\n",
       "      <td>0.321974</td>\n",
       "      <td>0.303632</td>\n",
       "      <td>0.351651</td>\n",
       "      <td>0.405474</td>\n",
       "      <td>0.495430</td>\n",
       "      <td>0.309546</td>\n",
       "      <td>0.355513</td>\n",
       "      <td>0.082500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 662 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CpG  X203203210100_R01C01  X203203210100_R02C01  \\\n",
       "0  cg00000165              0.101505              0.109506   \n",
       "1  cg00001209              0.874657              0.899091   \n",
       "2  cg00001364              0.809139              0.841104   \n",
       "3  cg00001582              0.040719              0.042991   \n",
       "4  cg00002920              0.415627              0.562619   \n",
       "\n",
       "   X203203210100_R03C01  X203203210100_R04C01  X203203210100_R05C01  \\\n",
       "0              0.141691              0.099335              0.119663   \n",
       "1              0.856218              0.896215              0.881654   \n",
       "2              0.768823              0.812510              0.767511   \n",
       "3              0.041190              0.047361              0.043020   \n",
       "4              0.492849              0.537996              0.646229   \n",
       "\n",
       "   X203203210100_R06C01  X203203210100_R01C02  X203203210100_R02C02  \\\n",
       "0              0.133722              0.160619              0.103879   \n",
       "1              0.919595              0.897742              0.895401   \n",
       "2              0.855345              0.825716              0.876175   \n",
       "3              0.043740              0.044908              0.037882   \n",
       "4              0.324161              0.240706              0.492182   \n",
       "\n",
       "   X203203210100_R03C02  ...  X203203210013_R03C01  X203203210013_R04C01  \\\n",
       "0              0.113632  ...              0.134059              0.111548   \n",
       "1              0.888192  ...              0.857043              0.826831   \n",
       "2              0.845755  ...              0.848691              0.865939   \n",
       "3              0.046859  ...              0.035247              0.034831   \n",
       "4              0.404285  ...              0.329255              0.315051   \n",
       "\n",
       "   X203203210013_R05C01  X203203210013_R06C01  X203203210013_R01C02  \\\n",
       "0              0.127344              0.084265              0.112711   \n",
       "1              0.866019              0.831738              0.852618   \n",
       "2              0.903710              0.920990              0.860893   \n",
       "3              0.039376              0.040148              0.040499   \n",
       "4              0.321974              0.303632              0.351651   \n",
       "\n",
       "   X203203210013_R02C02  X203203210013_R03C02  X203203210013_R04C02  \\\n",
       "0              0.098072              0.096010              0.117737   \n",
       "1              0.906398              0.884948              0.846944   \n",
       "2              0.895960              0.894021              0.896620   \n",
       "3              0.039100              0.040631              0.043228   \n",
       "4              0.405474              0.495430              0.309546   \n",
       "\n",
       "   X203203210013_R05C02  X203203210013_R06C02  \n",
       "0              0.093164              0.184909  \n",
       "1              0.924727              0.873856  \n",
       "2              0.906478              0.891837  \n",
       "3              0.045163              0.040598  \n",
       "4              0.355513              0.082500  \n",
       "\n",
       "[5 rows x 662 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_methylation_data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>GEO_ID</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X203203210100_R01C01</td>\n",
       "      <td>GSM5580349</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>50</td>\n",
       "      <td>Male</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X203203210100_R02C01</td>\n",
       "      <td>GSM5580350</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>44</td>\n",
       "      <td>Female</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X203203210100_R03C01</td>\n",
       "      <td>GSM5580351</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>52</td>\n",
       "      <td>Female</td>\n",
       "      <td>30-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X203203210100_R04C01</td>\n",
       "      <td>GSM5580352</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>62</td>\n",
       "      <td>Female</td>\n",
       "      <td>60-90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X203203210100_R05C01</td>\n",
       "      <td>GSM5580353</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>65</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Sample      GEO_ID   Tissue  Age     Sex Age Group\n",
       "0  X203203210100_R01C01  GSM5580349  Adipose   50    Male     30-60\n",
       "1  X203203210100_R02C01  GSM5580350  Adipose   44  Female     30-60\n",
       "2  X203203210100_R03C01  GSM5580351  Adipose   52  Female     30-60\n",
       "3  X203203210100_R04C01  GSM5580352  Adipose   62  Female     60-90\n",
       "4  X203203210100_R05C01  GSM5580353  Adipose   65    Male     60-90"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_metadata_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Heatmaps of methylation grouped by tissue were done in R and are available in folder \"Heatmaps\" withit the current directory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. Statistical Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of significant CpG sites: 103\n"
     ]
    }
   ],
   "source": [
    "# Perform t-test for each CpG site between age groups - All data\n",
    "\n",
    "# Import the necessary libraries\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import statsmodels.stats.multitest as multi\n",
    "\n",
    "# Load the merged data if needed\n",
    "human_merged_data = pd.read_csv('human_merged_data_184211.csv')\n",
    "\n",
    "# Create a list to store the results\n",
    "t_test_results = []\n",
    "\n",
    "# Iterate over each CpG site\n",
    "for cpg_site in human_merged_data.columns[6:]:\n",
    "    # Perform t-test between age groups\n",
    "    t_test_result = stats.ttest_ind(\n",
    "        human_merged_data[human_merged_data['Age Group'] == '0-30'][cpg_site],\n",
    "        human_merged_data[human_merged_data['Age Group'] == '30-60'][cpg_site]\n",
    "    )\n",
    "    # Store the results\n",
    "    t_test_results.append([cpg_site, t_test_result.pvalue])\n",
    "\n",
    "# Create a DataFrame from the results\n",
    "t_test_results_df = pd.DataFrame(t_test_results, columns=['CpG', 'P-Value'])\n",
    "\n",
    "# Perform multiple testing correction\n",
    "t_test_results_df['FDR'] = multi.fdrcorrection(t_test_results_df['P-Value'])[1]\n",
    "\n",
    "# Filter significant CpG sites\n",
    "significant_cpg_sites = t_test_results_df[t_test_results_df['FDR'] < 0.05]\n",
    "\n",
    "# Check the number of significant CpG sites\n",
    "print(f'Number of significant CpG sites: {len(significant_cpg_sites)}')\n",
    "\n",
    "# Save the results to a CSV file    \n",
    "t_test_results_df.to_csv('t_test_results_184211.csv', index=False)\n",
    "\n",
    "# Save the significant CpG sites to a CSV file\n",
    "significant_cpg_sites.to_csv('significant_cpg_sites_184211.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TO DO:\n",
    "# find the average of each Cpg in the full dataset\n",
    "# find the average of each Cpg in the top 100 CpGs dataset\n",
    "# compare the averages\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
