{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script serves to import the Clock data from the original supplementary file of Lu et al 2023 (sup_data_2_43587_2023_462_MOESM3_ESM.xlsx)into a pandas dataframe. The data is then used to plot the Clock data for the different species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Importing Pan-Mammal Clocks Results from Lu et al. 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the Clock data from the excel file (Lu et al. 2023) ###\n",
    "\n",
    "# Load epi clock results from excel\n",
    "clocks_excel_file = \"sup_data_2_43587_2023_462_MOESM3_ESM.xlsx\"\n",
    "\n",
    "# Specify the sheets to import \n",
    "sheets_to_import = list(range(15, 18)) #  sheets 16 to 18 (table S3.1-3.3)\n",
    "\n",
    "# Load the sheets, each as a separate data frame\n",
    "clocks_dfs = [pd.read_excel(clocks_excel_file, sheet_name=sheet, skiprows=[0, 1, 3]) for sheet in sheets_to_import] # skipping rows 1, 2, and 4 to remove the legends and intercept rows\n",
    " \n",
    "# Separate the data frames into individual data frames (clocks 1, 2, and 3)\n",
    "clock_1_df, clock_2_df, clock_3_df = clocks_dfs\n",
    "\n",
    "# Create the overlapped dfs for clocks 1, 2 and 3 \n",
    "overlap_1_2_df = pd.merge(clock_1_df, clock_2_df, on='Gene', how='inner')\n",
    "overlap_1_2_3_df = pd.merge(overlap_1_2_df, clock_3_df, on='Gene', how='inner')\n",
    "overlap_1_2_3_df = overlap_1_2_3_df.drop_duplicates(subset='Gene') # Drop duplicate columns resulting from the merge\n",
    "\n",
    "# Create the overlapped dfs for clocks 2 and 3\n",
    "overlap_2_3_df = pd.merge(clock_2_df, clock_3_df, on='Gene', how='inner')\n",
    "overlap_2_3_df = overlap_2_3_df.drop_duplicates(subset='Gene') # Drop duplicate columns resulting from the merge\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clock 1: (335, 14)\n",
      "Clock 2: (816, 14)\n",
      "Clock 3: (760, 14)\n",
      "Overlap 1 and 2: (435, 27)\n",
      "Overlap 2 and 3: (401, 27)\n",
      "Overlap 1, 2 and 3: (140, 40)\n"
     ]
    }
   ],
   "source": [
    "# Check size of data frames\n",
    "\n",
    "print(f\"Clock 1: {clock_1_df.shape}\")\n",
    "print(f\"Clock 2: {clock_2_df.shape}\")\n",
    "print(f\"Clock 3: {clock_3_df.shape}\")\n",
    "print(f\"Overlap 1 and 2: {overlap_1_2_df.shape}\")\n",
    "print(f\"Overlap 2 and 3: {overlap_2_3_df.shape}\") # should be 401\n",
    "print(f\"Overlap 1, 2 and 3: {overlap_1_2_3_df.shape}\") # should be 140"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>cg00249943</td>\n",
       "      <td>0.152786</td>\n",
       "      <td>5</td>\n",
       "      <td>136274770</td>\n",
       "      <td>TRPC7</td>\n",
       "      <td>TRPC7</td>\n",
       "      <td>57113</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>TRPC7_Exon</td>\n",
       "      <td>Exon (ENST00000513104.5/57113, exon 4 of 12)</td>\n",
       "      <td>13</td>\n",
       "      <td>56831383.0</td>\n",
       "      <td>Trpc7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>cg00250826</td>\n",
       "      <td>0.002770</td>\n",
       "      <td>7</td>\n",
       "      <td>128754670</td>\n",
       "      <td>CALU</td>\n",
       "      <td>CALU</td>\n",
       "      <td>813</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>CALU_Exon</td>\n",
       "      <td>Exon (ENST00000542996.6/813, exon 4 of 8)</td>\n",
       "      <td>6</td>\n",
       "      <td>29361701.0</td>\n",
       "      <td>Calu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>cg00292639</td>\n",
       "      <td>-0.129891</td>\n",
       "      <td>3</td>\n",
       "      <td>52393402</td>\n",
       "      <td>DNAH1</td>\n",
       "      <td>DNAH1</td>\n",
       "      <td>25981</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>DNAH1_Exon</td>\n",
       "      <td>Exon (ENST00000420323.6/25981, exon 66 of 78)</td>\n",
       "      <td>14</td>\n",
       "      <td>31265961.0</td>\n",
       "      <td>Dnah1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>cg00362836</td>\n",
       "      <td>0.048630</td>\n",
       "      <td>17</td>\n",
       "      <td>12473058</td>\n",
       "      <td>LINC00670</td>\n",
       "      <td>LINC00670</td>\n",
       "      <td>284034</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>LINC00670_Intergenic_upstream</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>11</td>\n",
       "      <td>65435676.0</td>\n",
       "      <td>Myocd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>cg00411555</td>\n",
       "      <td>-0.031366</td>\n",
       "      <td>17</td>\n",
       "      <td>44206588</td>\n",
       "      <td>UBTF</td>\n",
       "      <td>UBTF</td>\n",
       "      <td>7343</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>UBTF_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>11</td>\n",
       "      <td>102305986.0</td>\n",
       "      <td>Ubtf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock1 CHR    bp_hg38       Gene  Gene.hg19  \\\n",
       "0      2  cg00249943     0.152786   5  136274770      TRPC7      TRPC7   \n",
       "1      3  cg00250826     0.002770   7  128754670       CALU       CALU   \n",
       "2      4  cg00292639    -0.129891   3   52393402      DNAH1      DNAH1   \n",
       "3      5  cg00362836     0.048630  17   12473058  LINC00670  LINC00670   \n",
       "4      6  cg00411555    -0.031366  17   44206588       UBTF       UBTF   \n",
       "\n",
       "   ENTREZID                                conservationInMouse  \\\n",
       "0     57113       conserved gene and region in mouse and human   \n",
       "1       813  conserved gene but different region in human a...   \n",
       "2     25981       conserved gene and region in mouse and human   \n",
       "3    284034               mapped in human, not mapped in mouse   \n",
       "4      7343               mapped in human, not mapped in mouse   \n",
       "\n",
       "                    GeneRegionID  \\\n",
       "0                     TRPC7_Exon   \n",
       "1                      CALU_Exon   \n",
       "2                     DNAH1_Exon   \n",
       "3  LINC00670_Intergenic_upstream   \n",
       "4                  UBTF_threeUTR   \n",
       "\n",
       "                                      annotation CHR_mm10      bp_mm10  \\\n",
       "0   Exon (ENST00000513104.5/57113, exon 4 of 12)       13   56831383.0   \n",
       "1      Exon (ENST00000542996.6/813, exon 4 of 8)        6   29361701.0   \n",
       "2  Exon (ENST00000420323.6/25981, exon 66 of 78)       14   31265961.0   \n",
       "3                              Distal Intergenic       11   65435676.0   \n",
       "4                                         3' UTR       11  102305986.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0     Trpc7  \n",
       "1      Calu  \n",
       "2     Dnah1  \n",
       "3     Myocd  \n",
       "4      Ubtf  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clock_1_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock2</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>cg00020468</td>\n",
       "      <td>-0.002385</td>\n",
       "      <td>10</td>\n",
       "      <td>60070043</td>\n",
       "      <td>ANK3</td>\n",
       "      <td>ANK3</td>\n",
       "      <td>288</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>ANK3_Exon</td>\n",
       "      <td>Exon (ENST00000280772.7/288, exon 37 of 44)</td>\n",
       "      <td>10</td>\n",
       "      <td>69992409.0</td>\n",
       "      <td>Ank3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>cg00096922</td>\n",
       "      <td>-0.009445</td>\n",
       "      <td>7</td>\n",
       "      <td>97024873</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>1749</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>DLX5_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>6</td>\n",
       "      <td>6882127.0</td>\n",
       "      <td>Dlx5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>cg00098422</td>\n",
       "      <td>0.055805</td>\n",
       "      <td>1</td>\n",
       "      <td>224389815</td>\n",
       "      <td>WDR26</td>\n",
       "      <td>WDR26</td>\n",
       "      <td>80232</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>WDR26_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>1</td>\n",
       "      <td>181177786.0</td>\n",
       "      <td>Wdr26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>cg00106940</td>\n",
       "      <td>-0.059145</td>\n",
       "      <td>3</td>\n",
       "      <td>181712691</td>\n",
       "      <td>SOX2-OT</td>\n",
       "      <td>SOX2-OT</td>\n",
       "      <td>347689</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SOX2-OT_Exon</td>\n",
       "      <td>Exon (ENST00000325404.3/6657, exon 1 of 1)</td>\n",
       "      <td>3</td>\n",
       "      <td>34650752.0</td>\n",
       "      <td>Sox2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>cg00195533</td>\n",
       "      <td>-0.172659</td>\n",
       "      <td>7</td>\n",
       "      <td>128832266</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>2318</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>FLNC_Intron</td>\n",
       "      <td>Intron (ENST00000325888.12/2318, intron 1 of 47)</td>\n",
       "      <td>6</td>\n",
       "      <td>29435072.0</td>\n",
       "      <td>Flnc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock2 CHR    bp_hg38     Gene Gene.hg19  ENTREZID  \\\n",
       "0      2  cg00020468    -0.002385  10   60070043     ANK3      ANK3       288   \n",
       "1      3  cg00096922    -0.009445   7   97024873     DLX5      DLX5      1749   \n",
       "2      4  cg00098422     0.055805   1  224389815    WDR26     WDR26     80232   \n",
       "3      5  cg00106940    -0.059145   3  181712691  SOX2-OT   SOX2-OT    347689   \n",
       "4      6  cg00195533    -0.172659   7  128832266     FLNC      FLNC      2318   \n",
       "\n",
       "                                 conservationInMouse    GeneRegionID  \\\n",
       "0               mapped in human, not mapped in mouse       ANK3_Exon   \n",
       "1  conserved gene but different region in human a...    DLX5_fiveUTR   \n",
       "2       conserved gene and region in mouse and human  WDR26_threeUTR   \n",
       "3        mapped to different gene in human and mouse    SOX2-OT_Exon   \n",
       "4       conserved gene and region in mouse and human     FLNC_Intron   \n",
       "\n",
       "                                         annotation CHR_mm10      bp_mm10  \\\n",
       "0       Exon (ENST00000280772.7/288, exon 37 of 44)       10   69992409.0   \n",
       "1                                  Promoter (<=1kb)        6    6882127.0   \n",
       "2                                            3' UTR        1  181177786.0   \n",
       "3        Exon (ENST00000325404.3/6657, exon 1 of 1)        3   34650752.0   \n",
       "4  Intron (ENST00000325888.12/2318, intron 1 of 47)        6   29435072.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0      Ank3  \n",
       "1      Dlx5  \n",
       "2     Wdr26  \n",
       "3      Sox2  \n",
       "4      Flnc  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clock_2_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock3</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>cg00101675</td>\n",
       "      <td>-0.101354</td>\n",
       "      <td>14</td>\n",
       "      <td>63453806</td>\n",
       "      <td>PPP2R5E</td>\n",
       "      <td>PPP2R5E</td>\n",
       "      <td>5529</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>PPP2R5E_Exon</td>\n",
       "      <td>Exon (ENST00000337537.8/5529, exon 3 of 14)</td>\n",
       "      <td>12</td>\n",
       "      <td>75515719.0</td>\n",
       "      <td>Ppp2r5e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>cg06259996</td>\n",
       "      <td>0.045149</td>\n",
       "      <td>5</td>\n",
       "      <td>93585317</td>\n",
       "      <td>NR2F1-AS1</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>78198288.0</td>\n",
       "      <td>Nr2f1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.663272</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>cg15168457</td>\n",
       "      <td>0.766234</td>\n",
       "      <td>12</td>\n",
       "      <td>111516347</td>\n",
       "      <td>ATXN2</td>\n",
       "      <td>ATXN2</td>\n",
       "      <td>6311</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>ATXN2_Exon</td>\n",
       "      <td>Exon (ENST00000642389.2/6311, exon 10 of 27)</td>\n",
       "      <td>5</td>\n",
       "      <td>121781348.0</td>\n",
       "      <td>Atxn2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>cg20370622</td>\n",
       "      <td>0.549191</td>\n",
       "      <td>4</td>\n",
       "      <td>41748222</td>\n",
       "      <td>PHOX2B</td>\n",
       "      <td>PHOX2B</td>\n",
       "      <td>8929</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>PHOX2B_Promoter</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>5</td>\n",
       "      <td>67098553.0</td>\n",
       "      <td>Phox2b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index         var  beta_clock3 CHR    bp_hg38       Gene Gene.hg19  \\\n",
       "0      2  cg00101675    -0.101354  14   63453806    PPP2R5E   PPP2R5E   \n",
       "1      3  cg06259996     0.045149   5   93585317  NR2F1-AS1     NR2F1   \n",
       "2      4  cg08938156     0.663272   3  147409417  LOC440982      ZIC1   \n",
       "3      5  cg15168457     0.766234  12  111516347      ATXN2     ATXN2   \n",
       "4      6  cg20370622     0.549191   4   41748222     PHOX2B    PHOX2B   \n",
       "\n",
       "   ENTREZID                           conservationInMouse       GeneRegionID  \\\n",
       "0      5529  conserved gene and region in mouse and human       PPP2R5E_Exon   \n",
       "1    441094   mapped to different gene in human and mouse     NR2F1-AS1_Exon   \n",
       "2    440982   mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "3      6311  conserved gene and region in mouse and human         ATXN2_Exon   \n",
       "4      8929  conserved gene and region in mouse and human    PHOX2B_Promoter   \n",
       "\n",
       "                                     annotation CHR_mm10      bp_mm10  \\\n",
       "0   Exon (ENST00000337537.8/5529, exon 3 of 14)       12   75515719.0   \n",
       "1                              Promoter (<=1kb)       13   78198288.0   \n",
       "2                                        5' UTR        9   91365746.0   \n",
       "3  Exon (ENST00000642389.2/6311, exon 10 of 27)        5  121781348.0   \n",
       "4                              Promoter (<=1kb)        5   67098553.0   \n",
       "\n",
       "  Gene_mm10  \n",
       "0   Ppp2r5e  \n",
       "1     Nr2f1  \n",
       "2      Zic1  \n",
       "3     Atxn2  \n",
       "4    Phox2b  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clock_3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>cg00250826</td>\n",
       "      <td>0.002770</td>\n",
       "      <td>7</td>\n",
       "      <td>128754670</td>\n",
       "      <td>CALU</td>\n",
       "      <td>CALU</td>\n",
       "      <td>813</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>CALU_Exon</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>128754286</td>\n",
       "      <td>CALU</td>\n",
       "      <td>813</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>CALU_Exon</td>\n",
       "      <td>Exon (ENST00000535011.6/813, exon 3 of 6)</td>\n",
       "      <td>6</td>\n",
       "      <td>29361318.0</td>\n",
       "      <td>Calu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>cg00513357</td>\n",
       "      <td>0.091426</td>\n",
       "      <td>11</td>\n",
       "      <td>10806046</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>1982</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>EIF4G2_Exon</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>10806872</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>1982</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>EIF4G2_Exon</td>\n",
       "      <td>Exon (ENST00000526148.5/1982, exon 3 of 22)</td>\n",
       "      <td>7</td>\n",
       "      <td>111080951.0</td>\n",
       "      <td>Eif4g2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>cg00587168</td>\n",
       "      <td>0.055299</td>\n",
       "      <td>2</td>\n",
       "      <td>172093840</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>1745</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>DLX1_Intergenic_downstream</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>172085716</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>1745</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>DLX1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>71530032.0</td>\n",
       "      <td>Dlx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>11</td>\n",
       "      <td>cg00694357</td>\n",
       "      <td>-0.808471</td>\n",
       "      <td>1</td>\n",
       "      <td>155004171</td>\n",
       "      <td>ZBTB7B</td>\n",
       "      <td>ZBTB7B</td>\n",
       "      <td>51043</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>ZBTB7B_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>155007795</td>\n",
       "      <td>ZBTB7B</td>\n",
       "      <td>51043</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>ZBTB7B_Intron</td>\n",
       "      <td>Intron (ENST00000417934.6/51043, intron 1 of 4)</td>\n",
       "      <td>3</td>\n",
       "      <td>89388223.0</td>\n",
       "      <td>Zbtb7b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>cg01009870</td>\n",
       "      <td>0.101085</td>\n",
       "      <td>10</td>\n",
       "      <td>129869867</td>\n",
       "      <td>EBF3</td>\n",
       "      <td>EBF3</td>\n",
       "      <td>253738</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>EBF3_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>129873560</td>\n",
       "      <td>EBF3</td>\n",
       "      <td>253738</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>EBF3_Exon</td>\n",
       "      <td>Exon (ENST00000368648.7/253738, exon 8 of 16)</td>\n",
       "      <td>7</td>\n",
       "      <td>137231272.0</td>\n",
       "      <td>Ebf3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    index_x       var_x  beta_clock1 CHR_x  bp_hg38_x    Gene Gene.hg19_x  \\\n",
       "0         3  cg00250826     0.002770     7  128754670    CALU        CALU   \n",
       "2         7  cg00513357     0.091426    11   10806046  EIF4G2      EIF4G2   \n",
       "3         8  cg00587168     0.055299     2  172093840    DLX1        DLX1   \n",
       "5        11  cg00694357    -0.808471     1  155004171  ZBTB7B      ZBTB7B   \n",
       "14       14  cg01009870     0.101085    10  129869867    EBF3        EBF3   \n",
       "\n",
       "    ENTREZID_x                              conservationInMouse_x  \\\n",
       "0          813  conserved gene but different region in human a...   \n",
       "2         1982       conserved gene and region in mouse and human   \n",
       "3         1745        mapped to different gene in human and mouse   \n",
       "5        51043  conserved gene but different region in human a...   \n",
       "14      253738       conserved gene and region in mouse and human   \n",
       "\n",
       "                GeneRegionID_x  ... CHR    bp_hg38  Gene.hg19 ENTREZID  \\\n",
       "0                    CALU_Exon  ...   7  128754286       CALU      813   \n",
       "2                  EIF4G2_Exon  ...  11   10806872     EIF4G2     1982   \n",
       "3   DLX1_Intergenic_downstream  ...   2  172085716       DLX1     1745   \n",
       "5                ZBTB7B_Intron  ...   1  155007795     ZBTB7B    51043   \n",
       "14                 EBF3_Intron  ...  10  129873560       EBF3   253738   \n",
       "\n",
       "                             conservationInMouse   GeneRegionID  \\\n",
       "0           mapped in human, not mapped in mouse      CALU_Exon   \n",
       "2   conserved gene and region in mouse and human    EIF4G2_Exon   \n",
       "3    mapped to different gene in human and mouse      DLX1_Exon   \n",
       "5           mapped in human, not mapped in mouse  ZBTB7B_Intron   \n",
       "14  conserved gene and region in mouse and human      EBF3_Exon   \n",
       "\n",
       "                                         annotation CHR_mm10      bp_mm10  \\\n",
       "0         Exon (ENST00000535011.6/813, exon 3 of 6)        6   29361318.0   \n",
       "2       Exon (ENST00000526148.5/1982, exon 3 of 22)        7  111080951.0   \n",
       "3                                  Promoter (<=1kb)        2   71530032.0   \n",
       "5   Intron (ENST00000417934.6/51043, intron 1 of 4)        3   89388223.0   \n",
       "14    Exon (ENST00000368648.7/253738, exon 8 of 16)        7  137231272.0   \n",
       "\n",
       "   Gene_mm10  \n",
       "0       Calu  \n",
       "2     Eif4g2  \n",
       "3       Dlx1  \n",
       "5     Zbtb7b  \n",
       "14      Ebf3  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_1_2_3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock2</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR_y</th>\n",
       "      <th>bp_hg38_y</th>\n",
       "      <th>Gene.hg19_y</th>\n",
       "      <th>ENTREZID_y</th>\n",
       "      <th>conservationInMouse_y</th>\n",
       "      <th>GeneRegionID_y</th>\n",
       "      <th>annotation_y</th>\n",
       "      <th>CHR_mm10_y</th>\n",
       "      <th>bp_mm10_y</th>\n",
       "      <th>Gene_mm10_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>cg00096922</td>\n",
       "      <td>-0.009445</td>\n",
       "      <td>7</td>\n",
       "      <td>97024873</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>1749</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>DLX5_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>97024873</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>1749</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>DLX5_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>6</td>\n",
       "      <td>6882127.0</td>\n",
       "      <td>Dlx5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>cg00098422</td>\n",
       "      <td>0.055805</td>\n",
       "      <td>1</td>\n",
       "      <td>224389815</td>\n",
       "      <td>WDR26</td>\n",
       "      <td>WDR26</td>\n",
       "      <td>80232</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>WDR26_threeUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>224431495</td>\n",
       "      <td>WDR26</td>\n",
       "      <td>80232</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>WDR26_Exon</td>\n",
       "      <td>Exon (ENST00000414423.8/80232, exon 3 of 14)</td>\n",
       "      <td>1</td>\n",
       "      <td>181209021.0</td>\n",
       "      <td>Wdr26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>cg00106940</td>\n",
       "      <td>-0.059145</td>\n",
       "      <td>3</td>\n",
       "      <td>181712691</td>\n",
       "      <td>SOX2-OT</td>\n",
       "      <td>SOX2-OT</td>\n",
       "      <td>347689</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SOX2-OT_Exon</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>181712998</td>\n",
       "      <td>SOX2-OT</td>\n",
       "      <td>347689</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SOX2-OT_Exon</td>\n",
       "      <td>Exon (ENST00000325404.3/6657, exon 1 of 1)</td>\n",
       "      <td>3</td>\n",
       "      <td>34651059.0</td>\n",
       "      <td>Sox2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>cg00195533</td>\n",
       "      <td>-0.172659</td>\n",
       "      <td>7</td>\n",
       "      <td>128832266</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>2318</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>FLNC_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>128832266</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>2318</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>FLNC_Intron</td>\n",
       "      <td>Intron (ENST00000325888.12/2318, intron 1 of 47)</td>\n",
       "      <td>6</td>\n",
       "      <td>29435072.0</td>\n",
       "      <td>Flnc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>cg00250826</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>7</td>\n",
       "      <td>128754670</td>\n",
       "      <td>CALU</td>\n",
       "      <td>CALU</td>\n",
       "      <td>813</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>CALU_Exon</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>128754286</td>\n",
       "      <td>CALU</td>\n",
       "      <td>813</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>CALU_Exon</td>\n",
       "      <td>Exon (ENST00000535011.6/813, exon 3 of 6)</td>\n",
       "      <td>6</td>\n",
       "      <td>29361318.0</td>\n",
       "      <td>Calu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_x       var_x  beta_clock2 CHR_x  bp_hg38_x     Gene Gene.hg19_x  \\\n",
       "0        3  cg00096922    -0.009445     7   97024873     DLX5        DLX5   \n",
       "1        4  cg00098422     0.055805     1  224389815    WDR26       WDR26   \n",
       "2        5  cg00106940    -0.059145     3  181712691  SOX2-OT     SOX2-OT   \n",
       "3        6  cg00195533    -0.172659     7  128832266     FLNC        FLNC   \n",
       "4        7  cg00250826     0.007619     7  128754670     CALU        CALU   \n",
       "\n",
       "   ENTREZID_x                              conservationInMouse_x  \\\n",
       "0        1749  conserved gene but different region in human a...   \n",
       "1       80232       conserved gene and region in mouse and human   \n",
       "2      347689        mapped to different gene in human and mouse   \n",
       "3        2318       conserved gene and region in mouse and human   \n",
       "4         813  conserved gene but different region in human a...   \n",
       "\n",
       "   GeneRegionID_x  ... CHR_y  bp_hg38_y  Gene.hg19_y ENTREZID_y  \\\n",
       "0    DLX5_fiveUTR  ...     7   97024873         DLX5       1749   \n",
       "1  WDR26_threeUTR  ...     1  224431495        WDR26      80232   \n",
       "2    SOX2-OT_Exon  ...     3  181712998      SOX2-OT     347689   \n",
       "3     FLNC_Intron  ...     7  128832266         FLNC       2318   \n",
       "4       CALU_Exon  ...     7  128754286         CALU        813   \n",
       "\n",
       "                               conservationInMouse_y GeneRegionID_y  \\\n",
       "0  conserved gene but different region in human a...   DLX5_fiveUTR   \n",
       "1       conserved gene and region in mouse and human     WDR26_Exon   \n",
       "2        mapped to different gene in human and mouse   SOX2-OT_Exon   \n",
       "3       conserved gene and region in mouse and human    FLNC_Intron   \n",
       "4               mapped in human, not mapped in mouse      CALU_Exon   \n",
       "\n",
       "                                       annotation_y CHR_mm10_y    bp_mm10_y  \\\n",
       "0                                  Promoter (<=1kb)          6    6882127.0   \n",
       "1      Exon (ENST00000414423.8/80232, exon 3 of 14)          1  181209021.0   \n",
       "2        Exon (ENST00000325404.3/6657, exon 1 of 1)          3   34651059.0   \n",
       "3  Intron (ENST00000325888.12/2318, intron 1 of 47)          6   29435072.0   \n",
       "4         Exon (ENST00000535011.6/813, exon 3 of 6)          6   29361318.0   \n",
       "\n",
       "  Gene_mm10_y  \n",
       "0        Dlx5  \n",
       "1       Wdr26  \n",
       "2        Sox2  \n",
       "3        Flnc  \n",
       "4        Calu  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_2_3_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Ordering the df's by beta coefficient (measure of the strength and direction of the relationship between methylation and biological age)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These coefficients represent the effect of each CpG site on the predicted DNA methylation age based on the model. The values can be positive or negative and do not directly correspond to methylation levels (which are typically within the 0-1 range). The coefficients indicate the direction and strength of the relationship between each CpG site and the predicted DNA methylation age, as determined by the regression analysis \n",
    "\n",
    "The beta coefficients indicate the magnitude and direction of the effect that each predictor variable has on the outcome variable.\n",
    "\n",
    "For an epigenetic clock model like this one, where the outcome is a measure of biological age based on epigenetic markers, the beta coefficients in the beta_clock1 column represent the weights assigned to each CpG site (variables listed in the var column) in the model. These weights determine the contribution of each CpG site to the overall prediction of biological age.\n",
    "\n",
    "The sign of the beta coefficient indicates the direction of the relationship:\n",
    "\n",
    "    Positive beta: An increase in the methylation level of the CpG site is associated with an increase in the predicted biological age.\n",
    "    Negative beta: An increase in the methylation level of the CpG site is associated with a decrease in the predicted biological age.\n",
    "\n",
    "The magnitude of the beta coefficient reflects the strength of the association. Larger absolute values indicate a stronger impact on the predicted outcome.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order dataframes by descending beta coefficient values\n",
    "\n",
    "clock_1_df = clock_1_df.sort_values(by='beta_clock1', ascending=False)\n",
    "clock_2_df = clock_2_df.sort_values(by='beta_clock2', ascending=False)\n",
    "clock_3_df = clock_3_df.sort_values(by='beta_clock3', ascending=False)\n",
    "overlap_1_2_3_df = overlap_1_2_3_df.sort_values(by='beta_clock1', ascending=False)\n",
    "overlap_2_3_df = overlap_2_3_df.sort_values(by='beta_clock2', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>143</td>\n",
       "      <td>cg11728741</td>\n",
       "      <td>0.928874</td>\n",
       "      <td>8</td>\n",
       "      <td>41896469</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>286</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>ANK1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>8</td>\n",
       "      <td>22975146.0</td>\n",
       "      <td>Ank1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>297</td>\n",
       "      <td>cg24352905</td>\n",
       "      <td>0.853268</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>160</td>\n",
       "      <td>cg13058338</td>\n",
       "      <td>0.638660</td>\n",
       "      <td>12</td>\n",
       "      <td>54173598</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>SMUG1</td>\n",
       "      <td>23583</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>SMUG1_Intron</td>\n",
       "      <td>Intron (ENST00000635234.1/23583, intron 2 of 4)</td>\n",
       "      <td>15</td>\n",
       "      <td>103146996.0</td>\n",
       "      <td>Smug1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>316</td>\n",
       "      <td>cg26067250</td>\n",
       "      <td>0.510613</td>\n",
       "      <td>2</td>\n",
       "      <td>172085721</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>DLX1</td>\n",
       "      <td>1745</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>DLX1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>71530037.0</td>\n",
       "      <td>Dlx1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index         var  beta_clock1 CHR    bp_hg38       Gene Gene.hg19  \\\n",
       "141    143  cg11728741     0.928874   8   41896469       ANK1      ANK1   \n",
       "295    297  cg24352905     0.853268   5   77645452        OTP       OTP   \n",
       "96      98  cg08938156     0.843749   3  147409417  LOC440982      ZIC1   \n",
       "158    160  cg13058338     0.638660  12   54173598      SMUG1     SMUG1   \n",
       "314    316  cg26067250     0.510613   2  172085721       DLX1      DLX1   \n",
       "\n",
       "     ENTREZID                           conservationInMouse  \\\n",
       "141       286  conserved gene and region in mouse and human   \n",
       "295     23440  conserved gene and region in mouse and human   \n",
       "96     440982   mapped to different gene in human and mouse   \n",
       "158     23583   mapped to different gene in human and mouse   \n",
       "314      1745   mapped to different gene in human and mouse   \n",
       "\n",
       "          GeneRegionID                                       annotation  \\\n",
       "141          ANK1_Exon                                 Promoter (<=1kb)   \n",
       "295       OTP_Promoter                                 Promoter (5-6kb)   \n",
       "96   LOC440982_fiveUTR                                           5' UTR   \n",
       "158       SMUG1_Intron  Intron (ENST00000635234.1/23583, intron 2 of 4)   \n",
       "314          DLX1_Exon                                 Promoter (<=1kb)   \n",
       "\n",
       "    CHR_mm10      bp_mm10 Gene_mm10  \n",
       "141        8   22975146.0      Ank1  \n",
       "295       13   94869014.0       Otp  \n",
       "96         9   91365746.0      Zic1  \n",
       "158       15  103146996.0     Smug1  \n",
       "314        2   71530037.0      Dlx1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clock_1_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock2</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>142</td>\n",
       "      <td>cg05474883</td>\n",
       "      <td>0.952198</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>5</td>\n",
       "      <td>99962129.0</td>\n",
       "      <td>Hnrnpd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>272</td>\n",
       "      <td>cg09710440</td>\n",
       "      <td>0.877027</td>\n",
       "      <td>7</td>\n",
       "      <td>104328932</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>375612</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>LHFPL3_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>783</td>\n",
       "      <td>cg26512254</td>\n",
       "      <td>0.693060</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOTTIP</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>Intron (ENST00000421733.1/100316868, intron 1 ...</td>\n",
       "      <td>6</td>\n",
       "      <td>52260841.0</td>\n",
       "      <td>Hoxa13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>238</td>\n",
       "      <td>cg08681110</td>\n",
       "      <td>0.647719</td>\n",
       "      <td>6</td>\n",
       "      <td>113857322</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>4082</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>MARCKS_Promoter</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>10</td>\n",
       "      <td>37138948.0</td>\n",
       "      <td>Marcks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>253</td>\n",
       "      <td>cg09227056</td>\n",
       "      <td>0.615235</td>\n",
       "      <td>2</td>\n",
       "      <td>176075721</td>\n",
       "      <td>EVX2</td>\n",
       "      <td>EVX2</td>\n",
       "      <td>344191</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>EVX2_Intergenic_downstream</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>2</td>\n",
       "      <td>74651253.0</td>\n",
       "      <td>Evx2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index         var  beta_clock2 CHR    bp_hg38    Gene Gene.hg19  \\\n",
       "140    142  cg05474883     0.952198   4   82354926  HNRNPD    HNRNPD   \n",
       "270    272  cg09710440     0.877027   7  104328932  LHFPL3    LHFPL3   \n",
       "781    783  cg26512254     0.693060   7   27200144  HOTTIP    HOXA13   \n",
       "236    238  cg08681110     0.647719   6  113857322  MARCKS    MARCKS   \n",
       "251    253  cg09227056     0.615235   2  176075721    EVX2      EVX2   \n",
       "\n",
       "      ENTREZID                           conservationInMouse  \\\n",
       "140       3184  conserved gene and region in mouse and human   \n",
       "270     375612          mapped in human, not mapped in mouse   \n",
       "781  100316868   mapped to different gene in human and mouse   \n",
       "236       4082  conserved gene and region in mouse and human   \n",
       "251     344191  conserved gene and region in mouse and human   \n",
       "\n",
       "                   GeneRegionID  \\\n",
       "140             HNRNPD_threeUTR   \n",
       "270                 LHFPL3_Exon   \n",
       "781               HOTTIP_Intron   \n",
       "236             MARCKS_Promoter   \n",
       "251  EVX2_Intergenic_downstream   \n",
       "\n",
       "                                            annotation CHR_mm10     bp_mm10  \\\n",
       "140                                             3' UTR        5  99962129.0   \n",
       "270                                   Promoter (<=1kb)      NaN         NaN   \n",
       "781  Intron (ENST00000421733.1/100316868, intron 1 ...        6  52260841.0   \n",
       "236                                   Promoter (<=1kb)       10  37138948.0   \n",
       "251                                  Distal Intergenic        2  74651253.0   \n",
       "\n",
       "    Gene_mm10  \n",
       "140    Hnrnpd  \n",
       "270       NaN  \n",
       "781    Hoxa13  \n",
       "236    Marcks  \n",
       "251      Evx2  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clock_2_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var</th>\n",
       "      <th>beta_clock3</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>493</td>\n",
       "      <td>cg26512254</td>\n",
       "      <td>2.638299</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOTTIP</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>Intron (ENST00000421733.1/100316868, intron 1 ...</td>\n",
       "      <td>6</td>\n",
       "      <td>52260841.0</td>\n",
       "      <td>Hoxa13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>99</td>\n",
       "      <td>cg09710440</td>\n",
       "      <td>1.854689</td>\n",
       "      <td>7</td>\n",
       "      <td>104328932</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>LHFPL3</td>\n",
       "      <td>375612</td>\n",
       "      <td>mapped in human, not mapped in mouse</td>\n",
       "      <td>LHFPL3_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>34</td>\n",
       "      <td>cg18418719</td>\n",
       "      <td>1.806028</td>\n",
       "      <td>3</td>\n",
       "      <td>70972083</td>\n",
       "      <td>FOXP1</td>\n",
       "      <td>FOXP1</td>\n",
       "      <td>27086</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>FOXP1_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>6</td>\n",
       "      <td>98941067.0</td>\n",
       "      <td>Foxp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>134</td>\n",
       "      <td>cg24352905</td>\n",
       "      <td>1.738218</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>576</td>\n",
       "      <td>cg05474883</td>\n",
       "      <td>1.463495</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>5</td>\n",
       "      <td>99962129.0</td>\n",
       "      <td>Hnrnpd</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index         var  beta_clock3 CHR    bp_hg38    Gene Gene.hg19  \\\n",
       "491    493  cg26512254     2.638299   7   27200144  HOTTIP    HOXA13   \n",
       "97      99  cg09710440     1.854689   7  104328932  LHFPL3    LHFPL3   \n",
       "32      34  cg18418719     1.806028   3   70972083   FOXP1     FOXP1   \n",
       "132    134  cg24352905     1.738218   5   77645452     OTP       OTP   \n",
       "574    576  cg05474883     1.463495   4   82354926  HNRNPD    HNRNPD   \n",
       "\n",
       "      ENTREZID                           conservationInMouse     GeneRegionID  \\\n",
       "491  100316868   mapped to different gene in human and mouse    HOTTIP_Intron   \n",
       "97      375612          mapped in human, not mapped in mouse      LHFPL3_Exon   \n",
       "32       27086   mapped to different gene in human and mouse   FOXP1_threeUTR   \n",
       "132      23440  conserved gene and region in mouse and human     OTP_Promoter   \n",
       "574       3184  conserved gene and region in mouse and human  HNRNPD_threeUTR   \n",
       "\n",
       "                                            annotation CHR_mm10     bp_mm10  \\\n",
       "491  Intron (ENST00000421733.1/100316868, intron 1 ...        6  52260841.0   \n",
       "97                                    Promoter (<=1kb)      NaN         NaN   \n",
       "32                                              3' UTR        6  98941067.0   \n",
       "132                                   Promoter (5-6kb)       13  94869014.0   \n",
       "574                                             3' UTR        5  99962129.0   \n",
       "\n",
       "    Gene_mm10  \n",
       "491    Hoxa13  \n",
       "97        NaN  \n",
       "32      Foxp1  \n",
       "132       Otp  \n",
       "574    Hnrnpd  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clock_3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock1</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp_hg38</th>\n",
       "      <th>Gene.hg19</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>conservationInMouse</th>\n",
       "      <th>GeneRegionID</th>\n",
       "      <th>annotation</th>\n",
       "      <th>CHR_mm10</th>\n",
       "      <th>bp_mm10</th>\n",
       "      <th>Gene_mm10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>98</td>\n",
       "      <td>cg08938156</td>\n",
       "      <td>0.843749</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>LOC440982</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>147409417</td>\n",
       "      <td>ZIC1</td>\n",
       "      <td>440982</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>LOC440982_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>9</td>\n",
       "      <td>91365746.0</td>\n",
       "      <td>Zic1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>329</td>\n",
       "      <td>cg27201382</td>\n",
       "      <td>0.444645</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>109676257.0</td>\n",
       "      <td>Bdnf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>144</td>\n",
       "      <td>cg11904056</td>\n",
       "      <td>0.341791</td>\n",
       "      <td>5</td>\n",
       "      <td>93583204</td>\n",
       "      <td>NR2F1-AS1</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>93585317</td>\n",
       "      <td>NR2F1</td>\n",
       "      <td>441094</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>NR2F1-AS1_Exon</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>78198288.0</td>\n",
       "      <td>Nr2f1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>44</td>\n",
       "      <td>cg03820088</td>\n",
       "      <td>0.337644</td>\n",
       "      <td>7</td>\n",
       "      <td>23522229</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>29896</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>TRA2A_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>23522229</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>29896</td>\n",
       "      <td>conserved gene but different region in human a...</td>\n",
       "      <td>TRA2A_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>6</td>\n",
       "      <td>49252905.0</td>\n",
       "      <td>Tra2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>58</td>\n",
       "      <td>cg04998737</td>\n",
       "      <td>0.336275</td>\n",
       "      <td>5</td>\n",
       "      <td>77645403</td>\n",
       "      <td>OTP</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>77645452</td>\n",
       "      <td>OTP</td>\n",
       "      <td>23440</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>OTP_Promoter</td>\n",
       "      <td>Promoter (5-6kb)</td>\n",
       "      <td>13</td>\n",
       "      <td>94869014.0</td>\n",
       "      <td>Otp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index_x       var_x  beta_clock1 CHR_x  bp_hg38_x       Gene Gene.hg19_x  \\\n",
       "310       98  cg08938156     0.843749     3  147409417  LOC440982        ZIC1   \n",
       "899      329  cg27201382     0.444645    11   27720483       BDNF        BDNF   \n",
       "593      144  cg11904056     0.341791     5   93583204  NR2F1-AS1       NR2F1   \n",
       "171       44  cg03820088     0.337644     7   23522229      TRA2A       TRA2A   \n",
       "208       58  cg04998737     0.336275     5   77645403        OTP         OTP   \n",
       "\n",
       "     ENTREZID_x                              conservationInMouse_x  \\\n",
       "310      440982        mapped to different gene in human and mouse   \n",
       "899         627       conserved gene and region in mouse and human   \n",
       "593      441094        mapped to different gene in human and mouse   \n",
       "171       29896  conserved gene but different region in human a...   \n",
       "208       23440       conserved gene and region in mouse and human   \n",
       "\n",
       "        GeneRegionID_x  ... CHR    bp_hg38  Gene.hg19 ENTREZID  \\\n",
       "310  LOC440982_fiveUTR  ...   3  147409417       ZIC1   440982   \n",
       "899       BDNF_fiveUTR  ...  11   27720483       BDNF      627   \n",
       "593   NR2F1-AS1_Intron  ...   5   93585317      NR2F1   441094   \n",
       "171      TRA2A_fiveUTR  ...   7   23522229      TRA2A    29896   \n",
       "208       OTP_Promoter  ...   5   77645452        OTP    23440   \n",
       "\n",
       "                                   conservationInMouse       GeneRegionID  \\\n",
       "310        mapped to different gene in human and mouse  LOC440982_fiveUTR   \n",
       "899       conserved gene and region in mouse and human       BDNF_fiveUTR   \n",
       "593        mapped to different gene in human and mouse     NR2F1-AS1_Exon   \n",
       "171  conserved gene but different region in human a...      TRA2A_fiveUTR   \n",
       "208       conserved gene and region in mouse and human       OTP_Promoter   \n",
       "\n",
       "           annotation CHR_mm10      bp_mm10 Gene_mm10  \n",
       "310            5' UTR        9   91365746.0      Zic1  \n",
       "899  Promoter (<=1kb)        2  109676257.0      Bdnf  \n",
       "593  Promoter (<=1kb)       13   78198288.0     Nr2f1  \n",
       "171            5' UTR        6   49252905.0     Tra2a  \n",
       "208  Promoter (5-6kb)       13   94869014.0       Otp  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_1_2_3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>var_x</th>\n",
       "      <th>beta_clock2</th>\n",
       "      <th>CHR_x</th>\n",
       "      <th>bp_hg38_x</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene.hg19_x</th>\n",
       "      <th>ENTREZID_x</th>\n",
       "      <th>conservationInMouse_x</th>\n",
       "      <th>GeneRegionID_x</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR_y</th>\n",
       "      <th>bp_hg38_y</th>\n",
       "      <th>Gene.hg19_y</th>\n",
       "      <th>ENTREZID_y</th>\n",
       "      <th>conservationInMouse_y</th>\n",
       "      <th>GeneRegionID_y</th>\n",
       "      <th>annotation_y</th>\n",
       "      <th>CHR_mm10_y</th>\n",
       "      <th>bp_mm10_y</th>\n",
       "      <th>Gene_mm10_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>142</td>\n",
       "      <td>cg05474883</td>\n",
       "      <td>0.952198</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>82354926</td>\n",
       "      <td>HNRNPD</td>\n",
       "      <td>3184</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>HNRNPD_threeUTR</td>\n",
       "      <td>3' UTR</td>\n",
       "      <td>5</td>\n",
       "      <td>99962129.0</td>\n",
       "      <td>Hnrnpd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>783</td>\n",
       "      <td>cg26512254</td>\n",
       "      <td>0.693060</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOTTIP</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>27200144</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>100316868</td>\n",
       "      <td>mapped to different gene in human and mouse</td>\n",
       "      <td>HOTTIP_Intron</td>\n",
       "      <td>Intron (ENST00000421733.1/100316868, intron 1 ...</td>\n",
       "      <td>6</td>\n",
       "      <td>52260841.0</td>\n",
       "      <td>Hoxa13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>238</td>\n",
       "      <td>cg08681110</td>\n",
       "      <td>0.647719</td>\n",
       "      <td>6</td>\n",
       "      <td>113857322</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>4082</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>MARCKS_Promoter</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>113857322</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>4082</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>MARCKS_Promoter</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>10</td>\n",
       "      <td>37138948.0</td>\n",
       "      <td>Marcks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>803</td>\n",
       "      <td>cg27201382</td>\n",
       "      <td>0.523618</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>27720483</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>627</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>BDNF_fiveUTR</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>2</td>\n",
       "      <td>109676257.0</td>\n",
       "      <td>Bdnf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>599</td>\n",
       "      <td>cg19927064</td>\n",
       "      <td>0.499713</td>\n",
       "      <td>12</td>\n",
       "      <td>48818919</td>\n",
       "      <td>CACNB3</td>\n",
       "      <td>CACNB3</td>\n",
       "      <td>784</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>CACNB3_fiveUTR</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>48818919</td>\n",
       "      <td>CACNB3</td>\n",
       "      <td>784</td>\n",
       "      <td>conserved gene and region in mouse and human</td>\n",
       "      <td>CACNB3_fiveUTR</td>\n",
       "      <td>5' UTR</td>\n",
       "      <td>15</td>\n",
       "      <td>98635158.0</td>\n",
       "      <td>Cacnb3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_x       var_x  beta_clock2 CHR_x  bp_hg38_x    Gene Gene.hg19_x  \\\n",
       "341       142  cg05474883     0.952198     4   82354926  HNRNPD      HNRNPD   \n",
       "993       783  cg26512254     0.693060     7   27200144  HOTTIP      HOXA13   \n",
       "539       238  cg08681110     0.647719     6  113857322  MARCKS      MARCKS   \n",
       "1000      803  cg27201382     0.523618    11   27720483    BDNF        BDNF   \n",
       "899       599  cg19927064     0.499713    12   48818919  CACNB3      CACNB3   \n",
       "\n",
       "      ENTREZID_x                         conservationInMouse_x  \\\n",
       "341         3184  conserved gene and region in mouse and human   \n",
       "993    100316868   mapped to different gene in human and mouse   \n",
       "539         4082  conserved gene and region in mouse and human   \n",
       "1000         627  conserved gene and region in mouse and human   \n",
       "899          784  conserved gene and region in mouse and human   \n",
       "\n",
       "       GeneRegionID_x  ... CHR_y  bp_hg38_y  Gene.hg19_y ENTREZID_y  \\\n",
       "341   HNRNPD_threeUTR  ...     4   82354926       HNRNPD       3184   \n",
       "993     HOTTIP_Intron  ...     7   27200144       HOXA13  100316868   \n",
       "539   MARCKS_Promoter  ...     6  113857322       MARCKS       4082   \n",
       "1000     BDNF_fiveUTR  ...    11   27720483         BDNF        627   \n",
       "899    CACNB3_fiveUTR  ...    12   48818919       CACNB3        784   \n",
       "\n",
       "                             conservationInMouse_y   GeneRegionID_y  \\\n",
       "341   conserved gene and region in mouse and human  HNRNPD_threeUTR   \n",
       "993    mapped to different gene in human and mouse    HOTTIP_Intron   \n",
       "539   conserved gene and region in mouse and human  MARCKS_Promoter   \n",
       "1000  conserved gene and region in mouse and human     BDNF_fiveUTR   \n",
       "899   conserved gene and region in mouse and human   CACNB3_fiveUTR   \n",
       "\n",
       "                                           annotation_y CHR_mm10_y  \\\n",
       "341                                              3' UTR          5   \n",
       "993   Intron (ENST00000421733.1/100316868, intron 1 ...          6   \n",
       "539                                    Promoter (<=1kb)         10   \n",
       "1000                                   Promoter (<=1kb)          2   \n",
       "899                                              5' UTR         15   \n",
       "\n",
       "        bp_mm10_y Gene_mm10_y  \n",
       "341    99962129.0      Hnrnpd  \n",
       "993    52260841.0      Hoxa13  \n",
       "539    37138948.0      Marcks  \n",
       "1000  109676257.0        Bdnf  \n",
       "899    98635158.0      Cacnb3  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_2_3_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Check for missing values in the Gene and ENTREZ ID columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLOCK 1: Missing Genes: 0, Missing ENTREZ IDs: 0\n",
      "CLOCK 2: Missing Genes: 0, Missing ENTREZ IDs: 0\n",
      "CLOCK 3: Missing Genes: 0, Missing ENTREZ IDs: 0\n",
      "OOVERLAP 1, 2 and 3: Missing Genes: 0, Missing ENTREZsIDs: 0\n",
      "OOVERLAP 1, 2 and 3: Missing Genes: 0, Missing ENTREZsIDs: 0\n"
     ]
    }
   ],
   "source": [
    "# Create a list of the dataframes to loop through\n",
    "clock_dfs = [clock_1_df, clock_2_df, clock_3_df]\n",
    "overlap_dfs = [overlap_2_3_df, overlap_1_2_3_df]\n",
    "\n",
    "# Loop through the clock dataframes and check for missing genes and ENTREZIDs\n",
    "for i, df in enumerate(clock_dfs, start=1):\n",
    "    gene_missing = df['Gene'].isnull().sum()\n",
    "    entrezid_missing = df['ENTREZID'].isnull().sum()\n",
    "    print(f\"CLOCK {i}: Missing Genes: {gene_missing}, Missing ENTREZ IDs: {entrezid_missing}\")\n",
    "\n",
    "# Loop through the overlap dataframes and check for missing genes and ENTREZIDs\n",
    "for i, df in enumerate(overlap_dfs, start=2):\n",
    "    gene_missing = df['Gene'].isnull().sum()\n",
    "    entrezid_missing = df['ENTREZID_x'].isnull().sum()\n",
    "    print(f\"OOVERLAP 1, 2 and 3: Missing Genes: {gene_missing}, Missing ENTREZsIDs: {entrezid_missing}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Create and export individual csv files of each of the clocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export complete Clock 1 ordered by beta-values\n",
    "clock_1_df.to_csv('clock_1_ordered.csv', index=False)\n",
    "\n",
    "# Export complete Clock 2 ordered by beta-values\n",
    "clock_2_df.to_csv('clock_2_ordered.csv', index=False)\n",
    "\n",
    "# Export complete Clock 3 ordered by beta-values\n",
    "clock_3_df.to_csv('clock_3_ordered.csv', index=False)\n",
    "\n",
    "# Export complete Overlap 2 and 3 ordered by beta-values\n",
    "overlap_2_3_df.to_csv('overlap_2_3_ordered.csv', index=False)\n",
    "\n",
    "# Export complete Overlap 1, 2 and 3 ordered by beta-values\n",
    "overlap_1_2_3_df.to_csv('overlap_1_2_3_ordered.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Create and export csv files for the top 25 CpG sites with the highest beta coefficients for each clock"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 25 CpG's clock_1_df to csv\n",
    "clock_1_df.head(25).to_csv('top_25_cpgs_clock_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 25 CpG's clock_2_df to csv\n",
    "clock_2_df.head(25).to_csv('top_25_cpgs_clock_2.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 25 CpG's clock_3_df to csv\n",
    "clock_3_df.head(25).to_csv('top_25_cpgs_clock_3.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 25 CpG's overlap_1_2_3_df to csv\n",
    "overlap_1_2_3_df.head(25).to_csv('top_25_cpgs_overlap_1_2_3.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 25 CpG's overlap_2_3_df to csv\n",
    "overlap_2_3_df.head(25).to_csv('top_25_cpgs_overlap_2_3.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Create and export csv files for the top 100 CpG sites with the highest beta coefficients for each clock"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 100 CpG's clock_1_df to csv\n",
    "clock_1_df.head(100).to_csv('top_100_cpgs_clock_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 100 CpG's clock_2_df to csv\n",
    "clock_2_df.head(100).to_csv('top_100_cpgs_clock_2.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 100 CpG's clock_3_df to csv\n",
    "clock_3_df.head(100).to_csv('top_100_cpgs_clock_3.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 100 CpG's overlap_1_2_3_df to csv\n",
    "overlap_1_2_3_df.head(100).to_csv('top_100_cpgs_overlap_1_2_3.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the top 100 CpG's overlap_2_3_df to csv\n",
    "overlap_2_3_df.head(100).to_csv('top_100_cpgs_overlap_2_3.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
